---
title: "Final Pipeline, Linking GTEx SNPs to Phenotypes"
author: "Henry Young"
date: "2022-12-16"
output: html_document
---

###Please email me at harry.young@bristol.ac.uk with any questions about the pipeline
Please see instructions on github / in readme.md for details
## Assign your gene name, file directory and LD token
```{r}
#First line to clear out environment
rm(list=ls())
#1. Gene name list:
gene_name = "YOUR_GENE_NAME"
#e.g gene_name ="NFE2L2" ; "YOUR_GENE_NAME"

#2. setting the file directory to where you want it to go DRIVE:/YOUR_FILE_DIRECTORY/
file_directory = "DRIVE:/YOUR_FILE_DIRECTORY/"
#windows e.g: file_directory = "C:/Users/hy14913/OneDrive - University of Bristol/Desktop/May23_test"
#mac e.g: file_directory ="/Users/hy14913/Documents/Local_work_folder/JUL_TEST"

#3. LD link token setup
LD_token = "generic_token"
# 12 character token e.g:"abc1bc2bc3bc"
# "abc1bc2bc3bc"generic_token

try(if(gene_name=="YOUR_GENE_NAME") stop("Please enter your Gene Name"))
try(if(file_directory=="C:/Documents/Example directory") stop("Please enter your file_directory so the pipeline knows where to save"))
try(if(LD_token=="generic_token") stop("Please enter your LD token, acquired from https://ldlink.nci.nih.gov/?tab=apiaccess"))
```


##Prerequisite to load in :
```{r, results=FALSE, echo=FALSE}
# Check the packages you have and install the rest:
list_of_packages = c("circlize",
                     "tidyverse",
                     "viridis",
                     "ggpubr",
                     "devtools",
                     "LDlinkR",
                     "BiocManager",
                     "ggh4x",
                     "shiny",
                     "bslib",
                     "cowplot",
                     "plotly")

new_packages = list_of_packages[!(list_of_packages %in% installed.packages()[,"Package"])]
if(length(new_packages)>0){
  install.packages(new_packages)
}

p <- installed.packages()
is_bioc_installed = "BiocManager" %in% p

if(is_bioc_installed ==F){
  Sys.sleep(15)
  is_bioc_installed = "BiocManager" %in% p
}
```
Next, install the ones that aren't CRAN:
```{r}
# refresh list f installed packages and check if they're already there!
p <- installed.packages()
is_ieu_installed = "ieugwasr" %in% p
is_biomaRt_installed = "biomaRt" %in% p

library(BiocManager)
if(is_ieu_installed ==F){
remotes::install_github("mrcieu/ieugwasr", upgrade = "never")
}
if(is_biomaRt_installed ==F){
BiocManager::install(c("biomaRt"), update = FALSE) 
}
```



Now load in the final packages:
```{r}
library(circlize)
library(tidyverse)
library(viridis)
library(ggpubr)
library(devtools)
library(LDlinkR)
library(ggh4x)
library(shiny)
library(bslib)
library(cowplot)
library(ieugwasr)
library(biomaRt)
```


# Folder Setup:
sets up all folders and sub directories needed. 
```{r}
# Loop to print the names you'll make folders for
if(file.exists(paste0(file_directory, "/Trait and Tissue Groups"))){
  
}else{
  dir.create(paste0(file_directory, "/Trait and Tissue Groups"))
}
if(file.exists(paste0(file_directory, "/eGene"))){
 # Do nothing  
}else{
    dir.create(paste0(file_directory, "/eGene"))
}

#Moving files into appropriate folders for neatness
if(file.exists(paste0(file_directory,"/Full_trait_database.csv"))){
Full_trait_database = read_csv(paste0(file_directory,"/Full_trait_database.csv"))
file.remove(paste0(file_directory,"/Full_trait_database.csv"))
write_csv( Full_trait_database,paste0(file_directory,"/Trait and Tissue Groups/Full_trait_database.csv"))
}

if(file.exists(paste0(file_directory,"/Full_tissue_database.csv"))){
Full_tissue_database =  read_csv(paste0(file_directory,"/Full_tissue_database.csv"))
file.remove(paste0(file_directory,"/Full_tissue_database.csv"))
write_csv(Full_tissue_database ,paste0(file_directory,"/Trait and Tissue Groups/Full_tissue_database.csv"))
}

if(file.exists(paste0(file_directory,"/FINAL_Gtex_egene.csv"))){
egene =read_csv(paste0(file_directory,"/FINAL_Gtex_egene.csv"))
file.remove(paste0(file_directory,"/FINAL_Gtex_egene.csv"))
write_csv(egene ,paste0(file_directory,"/eGene/FINAL_Gtex_egene.csv"))
}

  #if it already exists, don't make it again
  if(file.exists(paste0(file_directory,"/" ,gene_name,"_test"))){
   print("file does not exist, creating a new folder")
  }else{
  dir.create(paste0(file_directory,"/" ,gene_name,"_test"))
  subfolder_names <- c("Circos","Ensembl","Gtex","LDMatrix",  "Gene_diagram_output", "Final_data", "Final_plots") 
  for (n in 1:length(subfolder_names)){
    folder =dir.create(paste0(file_directory,"/", gene_name,"_test/",subfolder_names[n]))
  }
}
```






# GTEx Data load in, cleaning and merging into final dataframes for plotting
## STEP 1: Organising each dataset
### STEP 1a: GTEx data:
Firstly, we read in the Gtex "SNPS associated with an expression change in the gene of interest" dataframe of interest and do some initial clean-up steps:

This is done automatically using their url SWAGGER entry to their database.
```{r Initial load and clean}
if(exists("Gtex_1")==T){
  print("df exists so moving on")
}else{
# Scraping version
Gtex_1 = read_tsv(paste0("https://gtexportal.org/rest/v1/association/singleTissueEqtl?format=tsv&geneSymbol=",gene_name,"&tissueSiteDetailId=Adipose_Subcutaneous%2CAdipose_Visceral_Omentum%2CAdrenal_Gland%2CArtery_Aorta%2CArtery_Coronary%2CArtery_Tibial%2CBladder%2CBrain_Amygdala%2CBrain_Anterior_cingulate_cortex_BA24%2CBrain_Caudate_basal_ganglia%2CBrain_Cerebellar_Hemisphere%2CBrain_Cerebellum%2CBrain_Cortex%2CBrain_Frontal_Cortex_BA9%2CBrain_Hippocampus%2CBrain_Hypothalamus%2CBrain_Nucleus_accumbens_basal_ganglia%2CBrain_Putamen_basal_ganglia%2CBrain_Spinal_cord_cervical_c-1%2CBrain_Substantia_nigra%2CBreast_Mammary_Tissue%2CCells_EBV-transformed_lymphocytes%2CCells_Cultured_fibroblasts%2CCells_Transformed_fibroblasts%2CCervix_Ectocervix%2CCervix_Endocervix%2CColon_Sigmoid%2CColon_Transverse%2CEsophagus_Gastroesophageal_Junction%2CEsophagus_Mucosa%2CEsophagus_Muscularis%2CFallopian_Tube%2CHeart_Atrial_Appendage%2CHeart_Left_Ventricle%2CKidney_Cortex%2CKidney_Medulla%2CLiver%2CLung%2CMinor_Salivary_Gland%2CMuscle_Skeletal%2CNerve_Tibial%2COvary%2CPancreas%2CPituitary%2CProstate%2CSkin_Not_Sun_Exposed_Suprapubic%2CSkin_Sun_Exposed_Lower_leg%2CSmall_Intestine_Terminal_Ileum%2CSpleen%2CStomach%2CTestis%2CThyroid%2CUterus%2CVagina%2CWhole_Blood&datasetId=gtex_v8"),
                 col_names = T
)
# reset names to look like web downloaded - saves me re-writing everything
Gtex_1 = Gtex_1 %>% 
  dplyr::rename(`Gencode Id` =gencodeId,
         `Gene Symbol` =geneSymbolUpper,
         `Gene Symbol Backup` =geneSymbol,
         `Variant Id` = variantId,
         `SNP Id` = snpId,
         `P-Value` = pValue,
         NES = nes,
         Tissue= tissueSiteDetailId)

df = Gtex_1
head(df)  # This function just shows us the top few rows of the dataframe to see what we're looking at
### Seprates out the variant ID into usable columns
df = df %>%
  separate(`Variant Id`, c("chr_gtex", "position_gtex", "a", "b", "build_gtex"), remove = F) 
  #remove = F keeps the original column too

#Finally quickly altering column name SNP ID to rsid for simplicity of code:
df = df%>%
  dplyr::rename(rsid = `SNP Id`)

head(df) 
#Also clear up any NA rsids that have snuck through
df = df[which(is.na(df$rsid)==F),]
# And "none"s  
df = df[which(df$rsid!="none"),]
df = df[which(df$rsid!="None"),]
}
```

So, now we have a Gtex table with a separated location and mutation column for easy use later on. IMPORTANT NOTE: a = reference allele and b = Effect or Alternate Allele. In other words, the change in expression NES occurs in individuals with the mutation at location x from a to b.

Modulus of NES setup - useful for ordering the dataframe later on: 
```{r Modulus NES setup}
df$`|NES|` = sqrt(df$NES*df$NES)
head(df)
tail(df) # Check that |NES| = NES in all cases 
```

Save this file as "cleaned" data:
```{r Save clean data}
write_excel_csv(df, paste0(file_directory,"/", gene_name,"_test/Gtex/", gene_name, "_Gtex_cleaned.csv"))
```

Then we need to isolate the unique SNPs for searching in the ieugwasR package:
```{r Unique SNPs from Gtex}
unique_Gtex_snps = unique(df$rsid) # Removes any duplicated rows
write.table(unique_Gtex_snps, paste0(file_directory,"/", gene_name,"_test/Gtex/unique_Gtex_snps.csv"), sep="\t", row.names=FALSE, col.names=FALSE) # saving the file
```


### Step 1b: Assess linkage disequilibrium
This is quite a complicated series of steps, going from the matrix file to clearing up linked snps. (This can also be done retrospectively but here we assign a value for LD to use later to trim down results). 


#### 1b - 2.1 Assess Linkage disequilibrium (r2)
First step is to search your list of snps (unique_Gtex_snps) into LDMatrix to get an interactive plot of linkage disequilibrium in your snp set. https://ldlink.nci.nih.gov/?tab=ldmatrix 

This is an automatable step but I think it is important to see the matrix plot in LD matrix WHEN YOU CAN (<300 online) , open your dataframe of unique SNPs in excel, copy and paste the column into the query box on LD matrix, tick all populations (or even just European) and click calculate.

From this you should get a nice matrix plot of all of of the linkage between your SNPs. You can save a copy of this into the LDMatrix/Plots folder and download the two csvs for D' and R2

I can add back in the dprime version but just used r2 as default for efficiencies sake

I calculate linkage at an r2 cutoff of 0.5 and 0.75, though 0.5 is what is used later on. If you want to modify this, let me know if you get stuck!

##### Get matrices of R2 
```{r Automated r2 START dataset }
if(exists("r2")){
  print("r2 already exists, moving on")
}else{
if(length(unique_Gtex_snps)<700){
r2 = LDmatrix(unique_Gtex_snps, pop = "EUR", r2d = "r2", token =LD_token)
write_csv(r2, paste0(file_directory,"/", gene_name,"_test/LDMatrix/",gene_name,"_r2.csv"))
# open up the .csv file from the website as tab-deliminated 
r2 = read_csv(paste0(file_directory,"/", gene_name,"_test/LDMatrix/",gene_name,"_r2.csv"))
head(r2)

  for(n in 1:ncol(r2)){#         LOOP to create a df of NA and values above 0.5 - CHANGE THIS CUTOFF TO WHAT YOU WANT
    if(n==1){ # First column is the rsid so skip it using 'next'
      next
    }
    else if(n==2){linked0.5 = r2[which(r2[,n]>=0.5),] # For the first column setup a new df to write to, called 'linked0.5'
      linked0.5 = linked0.5[,c(1,n)]
    }
    else{
      linked0.5_temp =r2[which(r2[,n]>=0.5),] # Each loop you are making a temporary file 'linked0.5_temp', 
      linked0.5_temp = linked0.5_temp[,c(1,n)] 
      linked0.5 = merge(linked0.5, linked0.5_temp, all.x=T, all.y=T) # Then merging to the original linked0.5 file
    }
  }

#Repeated for 0.75
  for(n in 1:ncol(r2)){#         LOOP to create a df of NA and values above 0.5 - CHANGE THIS CUTOFF TO WHAT YOU WANT
    if(n==1){ # First column is the rsid so skip it using 'next'
      next
    }
    else if(n==2){linked0.75 = r2[which(r2[,n]>=0.75),] # For the first column setup a new df to write to, called 'linked0.75'
      linked0.75 = linked0.75[,c(1,n)]
    }
    else{
      linked0.75_temp =r2[which(r2[,n]>=0.75),] # Each loop you are making a temporary file 'linked0.75_temp', 
      linked0.75_temp = linked0.75_temp[,c(1,n)] 
      linked0.75 = merge(linked0.75, linked0.75_temp, all.x=T, all.y=T) # Then merging to the original linked file
    }
  }

## Cleanup environment as it can get busy:
rm(linked0.5_temp, linked0.75_temp)

 ###Lines to convert numbers above 0.5 into rsID strings
linked_test0.5 =linked0.5 # duplicate the new dataframe
# This is a little complicated to understand so do read the description above
  for(n in 2:ncol(linked0.5)){
    linked_test0.5[,n] = with(linked_test0.5, ifelse(linked_test0.5[,n]>=0.5, linked_test0.5[,1], linked_test0.5[,n]))
  }
head(linked_test0.5)

# now we set up a r2ictionary so we can see which snps associate with which snps
rs_overlap_r20.5 = vector(mode="list", length = nrow(r2)) # set up a dictionary style list
names(rs_overlap_r20.5) = names(r2[,2:ncol(r2)]) # add in the names of the snps (rsids)

  for(n in 1:(ncol(linked_test0.5)-1)){
    rs_overlap_r20.5[[n]] = na.omit(linked_test0.5[,n+1]) # fill in the associated snps from the matrix columns
  }
test = na.omit(linked_test0.5[,1+1])


 ###Lines to convert numbers above 0.5 into rsID strings
linked_test0.75 =linked0.75 # duplicate the new dataframe
# This is a little complicated to understand so do read the description above
  for(n in 2:ncol(linked0.75)){
    linked_test0.75[,n] = with(linked_test0.75, ifelse(linked_test0.75[,n]>=0.75, linked_test0.75[,1], linked_test0.75[,n]))
  }

# now we set up a dictionary so we can see which snps associate with which snps
rs_overlap_r20.75 = vector(mode="list", length = nrow(r2)) # set up a dictionary style list
names(rs_overlap_r20.75) = names(r2[,2:ncol(r2)]) # add in the names of the snps (rsids)

  for(n in 1:(ncol(linked_test0.75)-1)){
    rs_overlap_r20.75[[n]] = na.omit(linked_test0.75[,n+1]) # fill in the associated snps from the matrix columns
  }
test = na.omit(linked_test0.75[,1+1])

# Saving for later

#Cleanup before saving the data
rm(linked_test0.5, linked_test0.75, linked0.5, linked0.75, temp, n,test,Gtex_1,r2)

##Save the environment
save.image(paste0(file_directory,"/", gene_name,"_test/LDMatrix/rs_overlap_ALL_",gene_name,".RData"))

}



if(length(unique_Gtex_snps)>=700){
 LIMITED_SNP_LIST = SNPclip(
   unique_Gtex_snps,
   pop = "EUR",
   r2_threshold = "0.5",
   maf_threshold = "0.01",
  token = LD_token )

LIMITED_SNP_LIST = LIMITED_SNP_LIST[grep("Variant kept.", LIMITED_SNP_LIST$Details),  ] # selects the ones that passed the test
LIMITED_SNP_LIST = as.data.frame(LIMITED_SNP_LIST[,1])
names(LIMITED_SNP_LIST) = "rsid"

write_csv(LIMITED_SNP_LIST, paste0(file_directory,"/", gene_name,"_test/LDMatrix/",gene_name, "_LIMITED_SNP_LIST.csv"))
unique_Gtex_snps = read_csv(paste0(file_directory,"/", gene_name,"_test/LDMatrix/",gene_name, "_LIMITED_SNP_LIST.csv"))
unique_Gtex_snps = unique(unique_Gtex_snps$rsid)
}
}
```


## STEP 2: PheWAS searching using ieugwasr 
### Searching snps in ieugwasr
searching IEU PheWAS database
This step takes a while, give it time, if it errors, chances are the servers are busy, try again or maybe try running in the evening.
```{r,Phewas searching }
unique_Gtex_snps = unique(df$rsid)
unique_Gtex_snps_df =as.data.frame(unique_Gtex_snps)
if(exists("completed_snps")){
  print(paste0("Already attempted run, ",length(completed_snps)," rsids searched - carrying on from where we left off"))
  m=(length(na.omit(completed_snps))%/%25+1)
  unique_Gtex_snps_df$test = !unique_Gtex_snps %in%completed_snps
  unique_Gtex_snps_df = unique_Gtex_snps_df %>% filter(test==T)
  unique_Gtex_snps = unique_Gtex_snps_df$unique_Gtex_snps
  for(n in 1:(length(na.omit(unique_Gtex_snps))%/%25 + 1)){
    test = na.omit(unique_Gtex_snps[(((n-1)*25)+1):(n*25)])
    print(paste0("Temporary dataframe ",n))
    Gtex_PHEWAS_2 = phewas(test, pval = 1e-04) # Adjustable p-value here #
    if(length(Gtex_PHEWAS_2)==12){
      if(nrow(Gtex_PHEWAS_2)>0){
          Gtex_PHEWAS = rbind(Gtex_PHEWAS, Gtex_PHEWAS_2)
          completed_snps2 = test
          completed_snps=c(completed_snps,completed_snps2)
          completed_snps=unique(completed_snps)
      }
    }else{break}  
  }
}else{
  for(n in 1:(length(na.omit(unique_Gtex_snps))%/%25 + 1)){
    test = na.omit(unique_Gtex_snps[(((n-1)*25)+1):(n*25)])
    
    if(n==1){
      print("First dataframe")
      Gtex_PHEWAS = phewas(test, pval = 1e-04) # Adjustable p-value here #
      completed_snps = test
      }
    else{
      print(paste0("Temporary dataframe ",n))
      Gtex_PHEWAS_2 = phewas(test, pval = 1e-04) # Adjustable p-value here #

      if(length(Gtex_PHEWAS_2)==12){
        if(nrow(Gtex_PHEWAS_2)>0){
          Gtex_PHEWAS = rbind(Gtex_PHEWAS, Gtex_PHEWAS_2)
          completed_snps2 = test
          completed_snps=c(completed_snps,completed_snps2)
          completed_snps=unique(completed_snps)
        }
      }else{break} 
    }
  }
}

if(nrow(Gtex_PHEWAS)==0){
  print("Houston, we have a problem. Re run this chunk using the play button in the top right of the box. If that still doesn't work, run line by line using ctrl-enter OR cmd-enter. If it still doesn't work then I'm afraid something has gone really wrong! email me at harry.young@bristol.ac.uk")
} else{
# Then save using the following line:
  Gtex_PHEWAS=unique(Gtex_PHEWAS)
  write_excel_csv(Gtex_PHEWAS, paste0(file_directory,"/", gene_name,"_test/Final_data/",gene_name,"_GtextoPHEWAS_RAW.csv"))
}


if(length(completed_snps)==length(unique(df$rsid))){
  print("All OK, moving on")
}else{
  print("run all from this chunk again")
}

```

## STEP 3: Merging the datasets

### GTEx (original, SNPs-Tissue) merge with PheWAS output (SNPs-Traits)

Firstly, load in these libraries and start to reorganize the Gtex-to-phewas RAW and original Gtex data to get more of a feel for what the data looks like:
```{r initial read in of dataframes and ordering step}

GtextoPHEWAS_RAW = read_csv(paste0(file_directory,"/", gene_name,"_test/Final_data/",gene_name,"_GtextoPHEWAS_RAW.csv")) # load in Phewas output
Gtex_cleaned = read_csv(paste0(file_directory,"/", gene_name,"_test/Gtex/", gene_name, "_Gtex_cleaned.csv")) #  load in Gtex input

GtextoPHEWAS_RAW = GtextoPHEWAS_RAW %>%
  rename(position_phewas = position, # Renaming columns to keep track of their original dataframe
         chr_phewas = chr)

#####Reordering PheWas and Gtex dataset - extra - to be removed #####
# NOT NECERSARY BUT MAKES IT EASIER TO LOOK AT !

 Gtex_cleaned = Gtex_cleaned[,c( "Gencode Id","Gene Symbol","rsid","a","b", "NES","|NES|","Tissue", "P-Value", "Variant Id","chr_gtex", "position_gtex", "build_gtex")]
 head(Gtex_cleaned)
 GtextoPHEWAS_RAW = GtextoPHEWAS_RAW[,c("rsid", "ea","nea","trait","beta", "p","se","n","eaf","chr_phewas","position_phewas", "id")] # "traitgroup",
 head(GtextoPHEWAS_RAW)
```

Next step is to merge these two dataframes: dyplr full_join or merge both work.

```{r Merging the gtex and phewas dataframes}

test = full_join(Gtex_cleaned, GtextoPHEWAS_RAW, by = "rsid")
test = test[,c("Gencode Id","Gene Symbol","rsid","a","b", "NES","|NES|","Tissue", "P-Value", "Variant Id","chr_gtex", "position_gtex", "build_gtex", "ea","nea","trait","beta", "p","se","n","eaf","chr_phewas","position_phewas", "id")]
head(test) 

```
The next step is cleaning up the merge and checking that it has worked as expected:
NOTE: the positions WILL NOT MATCH. phewas uses build 37 where gtex uses b38.

There will be some rows where it only has GTEx data, i.e. rsids that didn't come out with a result from PHEWAS searching.
To clean this, remove rows where there are NA's for multiple columns in PHEWAS section:

```{r cleaning up the merge step}
# test number of rows with NAs to see the issue:
test$na_count = apply(test, 1, function(x) sum(is.na(x)))
to_remove = test[which(test$na_count>5),]
head(to_remove) # any rows with lots of NAs in Phewas columns

## Remove rows with NA in these PHEWAS columns
test = test %>% drop_na(beta, eaf, chr_phewas, trait) 
write_excel_csv(test, paste0(file_directory,"/", gene_name,"_test/Gtex/", gene_name, "_merge_GTEX_PHEWAS.csv"))

```
### Correcting beta, ambiguous cases and adding groups:

Now that we have a merged table, we now need to filter out any ambiguous cases. These include A-T or C-G mutations as well as checking that the reference and effect alleles match for phewas/gtex and correcting the beta value if not.
### 3a. beta:
Firstly, Let's correct the beta value:
```{r setting up a corrected beta column}
df=test
#Sometimes the alleles slip into TRUE/FALSE columns so to correct that:
df$a = as.character(df$a)
df$b= as.character(df$b)
df$ea= as.character(df$ea)
df$nea = as.character(df$nea)

df$beta_cor = ifelse(df$ea == df$b & df$nea == df$a, df$beta,
                     ifelse(df$ea == df$a & df$nea == df$b, df$beta*-1, NA))

if ((sum(is.na(df$beta_cor)) == 0)==TRUE){
  print("There are no mutations that share an rsid that are completely different in Gtex and PheWAS")
} # if true then all good, if not you need to remove these cases as they are completely different.

df=df[which(is.na(df$beta_cor)==F),] # Removing cases where alleles are completely different - they are rare and just add confusion

```

### 3b. ambiguous cases (AT  CG):
To solve reverse vs sense strand issues
```{r setting up an ambiguous case column}


df$AT = ifelse(df$ea == "A" & df$nea == "T", "Yes", 
             ifelse(df$ea == "T" & df$nea == "A", "Yes",
                    ifelse(nchar(df$ea) != nchar(df$nea), "FS", "No")))

df$GC = ifelse(df$ea == "G" & df$nea == "C", "Yes", 
             ifelse(df$ea == "C" & df$nea == "G", "Yes",
                    ifelse(nchar(df$ea) != nchar(df$nea), "FS", "No")))


df$ambiguous = ifelse(df$AT =="Yes","AT/TA",
                     ifelse(df$GC =="Yes","GC/CG",
                            ifelse(df$GC =="FS" | df$AT=="FS", "FS", "No")))

number_of_ambiguous_cases = df[which(df$ambiguous != "No"),]

paste0("The percentage of ambiguous cases is ",nrow(number_of_ambiguous_cases)/nrow(df)*100,"%")  # Gives % of cases in df which are ambiguous.
head(df)

```
For the moment we are just labelling the ambiguous data, we can filter it out when plotting.

There is also the possibility that you might have an A-G mutant in Gtex and then find that the phewas is T-C. This might be due to a how the snp was found (reverse strand error), however, if that were the case we would pick it up in the beta_cor section as NA because the bases are 'completely different' So far, this hasn't happened but it is important to keep an eye out for and remove and beta_cor NA rows if needed.

The next step is to group the tissues and traits

### 3c-1. Trait grouping
I have gone through and categorised all the traits in the trait table from the open IEUOPENGWAS project. 

Feel free to go thorugh and alter categories and subcategories as they are HEAVILY subject to my own bias despite my care in reading on what each one is.

We need to assign those groups to the main dataframe:

```{r Assign Trait Groups}
# read in trait database
Full_trait_database = read_csv(paste0(file_directory,"/Trait and Tissue Groups/Full_trait_database.csv"))
Full_trait_database=unique(Full_trait_database)

df_backup=df

df = left_join(x=df_backup, y=Full_trait_database)

if(sum(is.na(df$traitgroup))==0){
  print("Merge succesful, no NA's")
} else{
  paste(df$trait[which(is.na(df$traitgroup)==T)], "needs to be added to Full_trait_database") # VERY UNLIKELY!
}

```


### 3c-2. Now for the Tissues:

This time the database won't change because Gtex has a finite number of tissues. Therefore open up Epidemiology/Trait and Tissue Groups/Full_tissue_database.csv:

```{r Assign Tissue Groups}
# load in tissue database
Full_tissue_database = read_csv(paste0(file_directory,"/Trait and Tissue Groups/Full_tissue_database.csv"))
df_backup = df
df = left_join(x=df_backup, y=Full_tissue_database)
```

Next it's going to be useful to have a simple binary column for positive and negative NES and beta values
```{r adding in binary NES and beta direction columns}
df$beta_direction = ifelse(df$beta_cor<0, "Negative",
							ifelse(df$beta_cor>0, "Positive", "No Change"))

df$NES_direction = ifelse(df$NES <0, "Negative",
							ifelse(df$NES>0, "Positive", "No Change"))

```


#### Reorganise for ease of looking at it
OK, final step is to reorganise this and spend a little time looking through the data making sure it all looks as expected. 

```{r Save Merged Dataset}

df = df[,c("Gene Symbol",
           "rsid",
           "a",
           "b",
           "NES",
           "|NES|",
           "Tissue",
           "Tissue_nice",
           "tissue_group",
           "tissue_group_secondary",
           "P-Value",
           "ea",
           "nea",
           "trait",
           "traitgroup",
           "Subcat",
           "Alt_Subcat",
           "beta_cor",
           "beta" , 
           "AT",
           "GC",
           "ambiguous",
           "p",
           "se",
           "n",
           "eaf",
           "chr_phewas",
           "position_phewas",
           "id",
           "Variant Id",
           "chr_gtex",
           "position_gtex",
           "build_gtex",
           "Gencode Id",
           "na_count",
           "beta_direction",
           "NES_direction")]

```

OK, looking good, we now have our new dataset, merged with the added columns for traitgroup and tissue group.


### 3d. More Data Cleaning
Firstly you need to clean up 'repeated traits'. This is annoying things like a comma in one name:

    e.g Forced vital capacity (FVC), Best measure" and "Forced vital capacity (FVC)  Best     measure"

These 2 appear as separate traits where they are essentially the same. So the following lines helps clean this up. I tend to look at the plots then come back to this code to add in more errors. But this serves as a list of ones I've found!

```{r Cleaning up duplicated traits}
# all of these work as replace the first argument with the second argument
df = df %>% 
  mutate(trait = ifelse(as.character(trait) == "Serum creatinine (eGFRcrea)", "Serum creatinine", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Heel bone mineral density (BMD) T-score, automated", "Heel bone mineral density (BMD) T-score automated",  as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "COPD/asthma/ILD related pneumonia or pneumonia derived septichaemia", "COPD/asthma related pneumonia or pneumonia derived septichaemia",  as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Estimated glomerular filtration rate (eGFR)", "Estimated glomerular filtration rate", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Forced expiratory volume in 1-second (FEV1), Best measure", "Forced expiratory volume in 1-second (FEV1)  Best measure", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Forced vital capacity (FVC), Best measure", "Forced vital capacity (FVC)  Best measure", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Heel bone mineral density", "Heel bone mineral density (BMD)", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Heel bone mineral density (BMD), T-score  automated", "Heel bone mineral density (BMD) T-score  automated", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Body mass index (BMI)", "Body mass index", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Apoliprotein B", "Apolipoprotein B", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "Apoliprotein A", "Apolipoprotein A", as.character(trait))) %>%
 mutate(trait = ifelse(as.character(trait) == "Forced expiratory volume in 1-second (FEV1), predicted percentage Respiratory Function","Forced expiratory volume in 1-second (FEV1)  predicted percentage Respiratory Function",  as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) =="Diastolic blood pressure  automated reading", "Diastolic blood pressure, automated reading", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="Systolic blood pressure  automated reading",  "Systolic blood pressure, automated reading", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="Blood clot  DVT  bronchitis  emphysema  asthma  rhinitis  eczema  allergy diagnosed by doctor: Emphysema/chronic bronchitis",  "Blood clot, DVT, bronchitis, emphysema, asthma, rhinitis, eczema, allergy diagnosed by doctor: Emphysema/chronic bronchitis", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) == "bipolar disorder", "Bipolar disorder", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "white blood cell count", "White blood cell count", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Neutrophill", "Neutrophil", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Mean corpuscular haemoglobin", "Mean corpuscular hemoglobin", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Haemoglobin concentration", "Hemoglobin concentration", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Low−density−lipoprotein cholesterol", "LDL cholesterol", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Neutrophill", "Neutrophil", as.character(trait))) %>%
  mutate(trait = ifelse(as.character(trait) == "Plateletcrit", "Platelet crit", as.character(trait))) %>% 
  mutate(trait = ifelse(as.character(trait) == "systolic blood pressure", "Systolic blood pressure", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="colorectal cancer",  "Colorectal Cancer", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="Crohn disease",  "Crohn's disease", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="Diagnoses - main ICD10: M70 Soft tissue disorders related to use, overuse and pressure",  "Diagnoses - main ICD10: M70 Soft tissue disorders related to use  overuse and pressure", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="diastolic blood pressure",  "Diastolic blood pressure", as.character(trait)))%>%
  mutate(trait = ifelse(as.character(trait) =="Diastolic blood pressure, automated reading",  "Diastolic blood pressure  automated reading", as.character(trait)))
  


#Follistatin
#follistatin
#Indolelactate
#Indolelacetate
#Forced expiratory volume in 1-second (FEV1)  predicted
#Forced expiratory volume in 1-second (FEV1)  predicted percentage

#Forced expiratory volume in 1-second (FEV1), predicted
#Forced expiratory volume in 1-second (FEV1), predicted percentage
# Multiple sclerosis
# multiple sclerosis
#myeloperoxidase
#Myeloperoxidase

#Obsessive Compulsive Disorder
#Obsessive-compulsive disorder

#P-Selectin
#P-selectin

#Prolactin
#prolactin
#Prolactin Receptor
#Prolactin receptor

#renin
#Renin

#resistin
#Resistin

#spondin 1
#Spondin-1

#Subjective well being
#Subjective well-being
# Loneliness  isolation
# Loneliness, isolation
# Long-standing illness  disability or infirmity
# Long-standing illness, disability or infirmity
# Major Depressive Disorder
# Major depressive disorder


FVC_test = df[which(df$trait =="Forced vital capacity (FVC)  Best measure"),]
FVC_test2 = df[which(df$trait =="Forced vital capacity (FVC), Best measure"),]
FVC_orginal = df[which(df$trait =="Forced vital capacity (FVC)  Best measure" |df$trait =="Forced vital capacity (FVC), Best measure"),]

head(FVC_test)

head(FVC_test2)

if(nrow(FVC_test2) == 0 & nrow(FVC_test) == nrow(FVC_orginal)){
  print("Correction for FVC worked!")
}

```

FVC_test2 should be empty and FVC_test  should have the same number of rows as FVC_original. You can modify these to include whichever one you want to test has worked OK. 

## Step 4: LD removal and other cleanup

The next 4 steps filter out all of the data we don't want:

### 4a. ukb-a is superfluous:
```{r Select and remove ukb-a}
df = df[grep("ukb-a", df$id, invert = TRUE), ]

```
### 4b. Next, we need to check that ukb-d and ukb-a aren't simply repeating each other in terms of a snps effect on a trait in most cases:
```{r test ukb-d/a overlap}
ukb_b = df[grep("ukb-b-", df$id),] #Select only these data
ukb_d = df[grep("ukb-d-", df$id),]
ukb_b = ukb_b[,c("rsid", "trait")]
ukb_d = ukb_d[,c("rsid", "trait")]#simplify to rsid and trait

test = inner_join(ukb_b,ukb_d)
nrow(test)
if(nrow(test)==0){
  print("There is no overlap between UKb-b and -d")
}
if(nrow(test)>0){
  print("There is overlap between UKb-b and -d, assess whether to remove or not!")
}
```
if nrow(test) is 0 then no overlap, if not you need to check what overlap there is and if it's problematic. (That may be ambiguous but you probably want to just filter out any duplicated results in this case. But that may not always be true so you've got to make that judgement call)

### 4c-2. - (3a). Ambiguous case removal: (may seem out of place but like LD, this was assessed earlier and not removed until now)
```{r Removal of any abiguous cases}
df = df[which(df$ambiguous == "No" | df$ambiguous == "FS"),]
#And finally save it:
write_csv(df, paste0(file_directory,"/", gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL.csv"))
```
### 4d. Ok now to deal with those pesky LD snps
This is a little more complicated to filter out retrospectively. Arguably I should have filtered out the linked snps prior to the search. The issue with this is that you may miss some associations. For example, if rs1 - kidney disease and rs1 is linked to rs 2 and rs 3, then we choose only to search rs3. rs3 isn't associated with anything so we miss the effect of a snp by filtering prior to the search.

It makes our lives harder here but hopefully it's worth it to have a better indication of snps' effect on disease.

The plan is as follows:
    
    Sort by p-value (phewas) from smallest to largest (most sig to least sig)
    For every unique rsid-trait combo,
    split the list of linked snps (LD0.75 or LD0.5) into a searchable string
    search the dataframe for each linked_rsid,
    if a linked_rsid is paired with the same trait as the original rsid-trait combo,
    Then keep only the highest p-value row.


```{r Removal of LD linked snps for r20.5}
#Data preamble##
unique_Gtex_snps= read_tsv(paste0(file_directory,"/", gene_name,"_test/Gtex/unique_Gtex_snps.csv"), col_names = FALSE) # read the original unique snps file
unique_Gtex_snps = unique_Gtex_snps$X1
if(length(unique_Gtex_snps)<700){
  rm(list=ls()[! ls() %in% c("file_directory","gene_name","LD_token","unique_Gtex_snps","completed_snps")])

load(paste0(file_directory,"/", gene_name,"_test/LDMatrix/rs_overlap_ALL_",gene_name,".RData"))  # specifically the dataframe with ambiguous data removed

df = read_csv(paste0(file_directory,"/", gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL.csv"))
backup =df

dfneg=backup[which(backup$beta_direction=="Negative"),]
backupneg=dfneg
dfneg = dplyr::select(dfneg, rsid, p, trait,beta_direction) # selects key columns for assigning linkage
dfneg= unique(dfneg) # remove all replicates created with tissue merge
dfneg$test ="unlinked"
arrange(dfneg, p)# Sort by p-value (phewas) from smallest to largest (most sig to least sig)
# unique rsid-trait combos 
rsid_trait_combos_neg = subset(dfneg, select = c(rsid, trait)) #simplify to two columns
rsid_trait_combos_neg = unique(rsid_trait_combos_neg) # pull out unique values

dfpos=backup[which(backup$beta_direction=="Positive"),]
backuppos=dfpos
dfpos = dplyr::select(dfpos, rsid, p, trait,beta_direction) # selects key columns for assigning linkage
dfpos= unique(dfpos) # remove all replicates created with tissue merge
dfpos$test ="unlinked"
# Sort by p-value (phewas) from smallest to largest (most sig to least sig)
arrange(dfpos, p)
# unique rsid-trait combos 
rsid_trait_combos_pos = subset(dfpos, select = c(rsid, trait)) #simplify to two columns
rsid_trait_combos_pos = unique(rsid_trait_combos_pos) # pull out unique values


####Negative rsid-traits########
for(n in 1:nrow(rsid_trait_combos_neg)){
  test_id = rsid_trait_combos_neg$rsid[n]
  test_trait = rsid_trait_combos_neg$trait[n]
  for(i in 1:length(rs_overlap_r20.5[[rsid_trait_combos_neg$rsid[n]]])){
    
    if(length(rs_overlap_r20.5[[rsid_trait_combos_neg$rsid[n]]])>1){ # if there is more than 1 linked_snp
      
      linked_id = rs_overlap_r20.5[[rsid_trait_combos_neg$rsid[n]]][i]# selects every rsid linked with the nth rsid 
      if(linked_id != test_id){# if the test and linked are not the same
        ##### If there is exactly one test rsid row#######
        if(nrow(dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),])==1){
          
          if(dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["test"]=="unlinked"& # has it been previously linked
             
             nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])>0){ # checking there are some linked_snps - nrow is not 0
            ##### If there is exactly one linked_rsid row######
            if(nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])==1){
              
             
              if(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["p"] >=
                    dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["p"]){ # If the linked_id-p value is higher/equal to, then mark it to be removed
                  
                dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["test"]="linked"# select all rows in dfneg that the linked snp and the test _trait is in, label linked
                }
              else{# else mark the test_id it to be removed
                  dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["test"]="linked"
                }
          }
            ##### if there is more than one linked_rsid low#########
              else if(nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])>1){
                  
                minpl = min(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait ),]["p"])
                if(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait & dfneg$p == minpl),]["p"] >=
                      dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["p"]){
                    
                  dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["test"]="linked"# select all rows in dfneg that the linked snp and the test _trait is in, label linked
                  }
                else{
                    dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["test"]="linked"
                  }
            }
          }
        }
        ##### If there is more than one test rsid row#####
        else if(nrow(dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),])>1){
          
          minpt = min(dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait ),]["p"])
            if(dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait & dfneg$p == minpt),]["test"]=="unlinked"& # has it been previously linked
                nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])>0){
              #### if there is exactly one linked_rsid row####
              if(nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])==1){
                
                if(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["p"] >=
                    dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait & dfneg$p == minpt),]["p"]){
                  
                dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["test"]="linked"# select all rows in dfneg that the linked snp and the test _trait is in, label linked
                  }
                else{
                  dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["test"]="linked"
                  }
              }
              ####else if there is more than one linked_rsid row
              else if(nrow(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),])>1){
                
                minpl = min(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait ),]["p"])
                if(dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait & dfneg$p == minpl),]["p"] >=
                    dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait& dfneg$p == minpt),]["p"]){
                  
                dfneg[which(dfneg$rsid == linked_id & dfneg$trait == test_trait),]["test"]="linked"
                  }
                else{
                  dfneg[which(dfneg$rsid == test_id & dfneg$trait == test_trait),]["test"]="linked"
                  }
              }
          }
        }
      }
    }
  }
}

test2neg = full_join(dfneg, backupneg, by = c("rsid","p","trait")) # merging back with the original table

testneg = test2neg[which(test2neg$test=="unlinked"),]

###Positive rsid-traits####


for(n in 1:nrow(rsid_trait_combos_pos)){
  
  test_id = rsid_trait_combos_pos$rsid[n]
  test_trait = rsid_trait_combos_pos$trait[n]
  for(i in 1:length(rs_overlap_r20.5[[rsid_trait_combos_pos$rsid[n]]])){
    
    if(length(rs_overlap_r20.5[[rsid_trait_combos_pos$rsid[n]]])>1){ # if there is more than 1 linked_snp
      
      linked_id = rs_overlap_r20.5[[rsid_trait_combos_pos$rsid[n]]][i]# selects every rsid linked with the nth rsid 
      if(linked_id != test_id){# if the test and linked are not the same
        ##### If there is exactly one test rsid row#######
        if(nrow(dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),])==1){
          
          if(dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["test"]=="unlinked"& # has it been previously linked
             
             nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])>0){ # checking there are some linked_snps - nrow is not 0
            ##### If there is exactly one linked_rsid row######
            if(nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])==1){
                
              if(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["p"] >=
                    dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["p"]){ # If the linked_id-p value is higher/equal to, then mark it to be removed
                  
                dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["test"]="linked"# select all rows in dfpos that the linked snp and the test _trait is in, label linked
                }
              else{# else mark the test_id it to be removed
                  dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["test"]="linked"
                }
          }
            ##### if there is more than one linked_rsid low#########
              else if(nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])>1){
                  
                minpl = min(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait ),]["p"])
                if(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait & dfpos$p == minpl),]["p"] >=
                      dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["p"]){
                    
                  dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["test"]="linked"# select all rows in dfpos that the linked snp and the test _trait is in, label linked
                  }
                else{
                    dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["test"]="linked"
                  }
            }
          }
        }
        ##### If there is more than one test rsid row#####
        else if(nrow(dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),])>1){
          
          minpt = min(dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait ),]["p"])
            if(dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait & dfpos$p == minpt),]["test"]=="unlinked"& # has it been previously linked
                nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])>0){
              #### if there is exactly one linked_rsid row####
              if(nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])==1){
                
                if(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["p"] >
                    dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait & dfpos$p == minpt),]["p"]){
                  
                dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["test"]="linked"# select all rows in dfpos that the linked snp and the test _trait is in, label linked
                  }
                else{
                  dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["test"]="linked"
                  }
              }
              ####else if there is more than one linked_rsid row
              else if(nrow(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),])>1){
                
                minpl = min(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait ),]["p"])
                if(dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait & dfpos$p == minpl),]["p"] >=
                    dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait& dfpos$p == minpt),]["p"]){
                  
                dfpos[which(dfpos$rsid == linked_id & dfpos$trait == test_trait),]["test"]="linked"
                  }
                else{
                  dfpos[which(dfpos$rsid == test_id & dfpos$trait == test_trait),]["test"]="linked"
                  }
              }
          }
        }
      }
    }
  }
}

test2pos = full_join(dfpos, backuppos, by = c("rsid","p","trait"))
# Saving it 
test2=rbind(test2neg,test2pos)
test2 = test2 %>% rename(beta_direction = beta_direction.y)
test2 = test2 %>% dplyr::select(-c("beta_direction.x"))
write_csv(test2, paste0(file_directory,"/",gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL_linkagetempr20.5.csv"))# full dataframe but with the "test" column
test = test2[which(test2$test=="unlinked"),]
df=test
} else{
  df = read_csv(paste0(file_directory,"/",gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL.csv"))
  df$test ="filler"
}

```
###4e Finalisation
```{r}
final_df = df 
# setting up a logp and directional logp column
final_df$p= ifelse(final_df$p==0, 1e-300,final_df$p)# if p is 0 - a strange error with p being so small it defaults to 0 - replace with a very small value so it doesn't break our plots
final_df$logp=-log10(final_df$p)
final_df$logp_DIR =1
final_df$logp_DIR = if_else(final_df$NES_direction =="Negative", -1,1)
final_df$logp_DIR = final_df$logp*final_df$logp_DIR

# Also, Other catergory needs sorting properly
final_df$traitgroup =as.character(final_df$traitgroup)
final_df$traitgroup = ifelse(final_df$Subcat=="Other Gene Expression Change", "Other Gene Expression Change", final_df$traitgroup)
final_df=unique(final_df)

write_csv(final_df,paste0(file_directory,"/", gene_name, "_test/Final_data/final_df.csv"))
if(file.exists(paste0(file_directory,"/",gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL.csv"))){
file.remove(paste0(file_directory,"/",gene_name,"_test/Gtex/",gene_name,"_gtexPHEWAS_no_ukaORAmbig_FINAL.csv"))
}


```

# Step 5: Plotting the data so far:
This section is a little less rigid and I would encourage you to think about what kind of things you might be able to pull from the dataset.

## Now for the actual plotting:
NOTE: I'm not going to go line by line for the ggplotting functions in all of the code below as it's all explained very nicely in the gg cheatsheet available online. 

For this section I also have made new file directories for each subheading of plots for easier browsing later. However, I have left these out of the code so you can just click through without the complicated sub-folder system. 

Circos Figure:
Position_Gtex_P
Tissue_trait_overlap
trait_position
+ curved gene diagrams:
Curved_gene_diagram_P_FULL_Colour
Curved_gene_diagram_RED_colour

## our df plotting
Initial sort of traitgroups into a custom order and colour scheme (to be used later)
```{r Reading in the FINAL df}
# Data load in and sort 
df = read_csv(paste0(file_directory,"/", gene_name, "_test/Final_data/final_df.csv"))

df = df %>% dplyr::select(-Alt_Subcat)# remove alt_subcat at least for the moment, as it seems to making duplication events
df = unique(df)

# removing duplicated raw values from more recent irnt
df = df %>% tidyr::separate(id,into=c("remove1","remove2","remove3","irntvsraw" ), remove=F) %>%
  filter((irntvsraw !="raw") %>%replace_na(TRUE)) %>%
  dplyr::select(-remove1,
                -remove2,
                -remove3,
                -irntvsraw)

df$beta_direction = factor(df$beta_direction, levels=c("Positive", "Negative"))
original_gtex =read_csv(paste0(file_directory,"/", gene_name,"_test/Gtex/", gene_name, "_Gtex_cleaned.csv"))
Full_tissue_database = read_csv(paste0(file_directory,"/Trait and Tissue Groups/Full_tissue_database.csv"))
original_gtex = left_join(original_gtex,Full_tissue_database)
trait_order =c("Body Fat/Size",
                                  "Diet",
                                  "Metabolism (inc. Diabetes)",
                                  "Tumours (Benign/Cancer)",
                                  "Atrophy/Dystrophy", 
                                  "Congenital Defects", 
                                  "Cardiovascular Traits",
                                  "Blood/Vascular Traits",
                                  "Inflammation" , 
                                  "Respiratory",
                                  "Renal/Urinary",
                                  "Physical Activity",
                                 
                                  "Digestive Traits",
                                  "Nervous System/Brain",
                                  "Bone Traits (inc. Dental)",
                                  "Muscle/Tendon Traits",
                                  "Physical Injury",
                                  "Alcohol-related Disease",
                                  "Life Span",
                                  "Female Specific Traits",
                                  "Male Specific Traits",
                                  "Disease (unknown origin)",
                                  "Other Gene Expression Change",
                                  "Other/Unspecified"
            )
df$traitgroup = factor(df$traitgroup, levels =trait_order)

# INCOMPLETE - From later, moved back for completions sake?
grid_col = c(`Body Fat/Size` = "#0DF485",
             `Diet` = "#39FFC0",
            `Metabolism (inc. Diabetes)` ="#32DCC1",
             `Tumours (Benign/Cancer)` ="#32CCDC" ,
             `Atrophy/Dystrophy` = "#0DB5F4",
            `Congenital Defects` = "#3277DC",
             `Cardiovascular Traits` = "#4532DC",
             `Blood/Vascular Traits` ="#6C39FF",
             Inflammation ="#9A32DC", #  or #AA98A970
            `Respiratory` ="#C432DC",
             `Renal/Urinary` = "#DC32D0",
            `Physical Activity` = "#DC324D",
             `Digestive Traits` ="#DC4132",
            `Nervous System/Brain` = "#DC5D32",
             `Bone Traits (inc. Dental)` = "#DC8732",
            `Muscle/Tendon Traits` = "#DCA732",
            `Physical Injury` = "#DCD232",
            `Alcohol-related Disease` ="#BCDC32",
             `Life Span` ="#92DC32",
             `Female Specific Traits` = "#43DC32",
            `Male Specific Traits` = "#00f556",
            `Disease (unknown origin)` ="#00f577",
            `Other/Unspecified` ="#C9C7BD",
            `Other Gene Expression Change` ="#78C47D",
             # Tissues:
             Blood = "#CC5500",
             `Cultured Cells` = "#00FFFF",
             Digestive ="#A7C7E7",
             Endocrine = "#FF4433",
             Fat = "#E0115F",
             `Female Reproductive Tissue` = "#0C5F7D",
             Heart = "#80461B",
             Lymphatic ="#023020",
             `Male Specific Tissue` ="#FFA500",
             Mammary = "#E9846E",
             Muscle = "#FFBF00",
             Nervous = "#90EE90",
             `Respiratory Tissue` = "#D673C9",
             Skin = "#00A36C",
            Urniary = "darkorchid4",
             Vascular ="#630330"
             
)

```
#### A1 - TissueGROUP-rsid_count_OVERVIEW ORIGINAL GTEX
This plot shows the overview of our SNPs in each tissuegroup from the Original GTEx dataframe. (i.e after phewas search and LD removal but then filtered to remove duplicated tissuegroups from the trait matching process)
```{r}
list_of_tissuegroups = unique(Full_tissue_database$tissue_group)
list_of_tissuegroups = sort(list_of_tissuegroups)

original_gtex$NES_direction =ifelse(original_gtex$NES>0, "Positive", "Negative")

test = original_gtex %>% group_by(tissue_group, NES_direction) %>% summarise(count = length(tissue_group[!is.na(tissue_group)])) # frequency by direction


original_gtex$tissue_group = ifelse(original_gtex$tissue_group=="Male_Specific_Tissue", "Male Specific Tissue", original_gtex$tissue_group)

# loop to add in the missing tissuegroups
for(tissuegroup in list_of_tissuegroups){
  for(n in 1:nrow(test)){
    if(nrow(test[which(test$tissue_group==tissuegroup),])==0){
    test[nrow(test)+1,]= test[nrow(test)+1,] # make a new row
    test$tissue_group[nrow(test)] =tissuegroup
    test$count[nrow(test)] = 0
    test$NES_direction[nrow(test)] = "Positive" # set it to either, doesn't matter as value is 0
    }
  }
}


test$NES_direction = factor(test$NES_direction, levels=c("Positive", "Negative"))
test= test[ which(is.na(test$tissue_group)==F),]

A1SNPcountperTissuegroup = ggplot(test, aes(x=tissue_group,y=count, fill=NES_direction)) + 
  geom_col()+ # bar chart counting how many times a tissue appears
  scale_fill_viridis_d(begin=1, end=0, name=paste0("Expression Change\n          (NES)")) + 
  labs(title= "Number of SNPs per Tissue from GTEx (before PheWAS and LD removal)" , x= "Tissue Group", y=paste("Number of SNPs"))+ #labels
  theme(axis.text.x = element_text(angle = 45, hjust =1),
        legend.position = "top",
         panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
    scale_x_discrete(expand = c(0.04, 0)) +
    scale_y_continuous(expand = c(0, 0)) 

A1SNPcountperTissuegroup

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/A1-Number of SNPs per Tissue from GTEx .pdf"), width= 150, height=100, unit="mm")

```
#### A2 - number_of_SNP-trait_associations_per_tissue_group after LD and pipeline run
This plot shows the overview of our SNPs in each Tissue Group from the final df. (i.e after PheWAS search and LD removal but then filtered to remove duplicated Tissue Groups from the trait matching process)
```{r}
LD_Gtex = df%>%
  dplyr::select(Tissue, tissue_group, rsid, `P-Value`, NES, position_gtex, chr_gtex, a, b, NES_direction)
LD_Gtex = unique(LD_Gtex)# Removes duplicated tissues

list_of_tissuegroups = unique(Full_tissue_database$tissue_group)
list_of_tissuegroups = sort(list_of_tissuegroups)

test = LD_Gtex %>% group_by(tissue_group, NES_direction) %>% summarise(count = length(tissue_group[!is.na(tissue_group)])) # frequency by direction

LD_Gtex$tissue_group = ifelse(LD_Gtex$tissue_group=="Male_Specific_Tissue", "Male Specific Tissue", LD_Gtex$tissue_group)

# loop to add in the missing tissuegroups
for(tissuegroup in list_of_tissuegroups){
  for(n in 1:nrow(test)){
    if(nrow(test[which(test$tissue_group==tissuegroup),])==0){
    test[nrow(test)+1,]= test[nrow(test)+1,] # make a new row
    test$tissue_group[nrow(test)] =tissuegroup
    test$count[nrow(test)] = 0
    test$NES_direction[nrow(test)] = "Positive" # set it to either, doesn't matter as value is 0
    }
  }
}

test$NES_direction = factor(test$NES_direction, levels=c("Positive", "Negative"))
test= test[ which(is.na(test$tissue_group)==F),]

A2SNPcountperTissuegroup = ggplot(test, aes(x=tissue_group, y=count, fill=NES_direction)) + 
  geom_col()+ # bar chart counting how many times a tissue appears
  #ylim(0,0.5)+
  scale_fill_viridis_d(begin=1, end=0, name=paste0("Expression Change\n          (NES)")) + # I like this palette but you can remove it or change it 
  labs(title= "Number of SNPs (after PheWAS search and LD removal)" , x= "Tissue Group", y=paste("Number of SNPs"))+ #labels
  theme(axis.text.x = element_text(angle = 45, hjust =1),
        legend.position = "top",
         panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
    scale_x_discrete(expand = c(0.04, 0)) +
    scale_y_continuous(expand = c(0, 0)) 
  # setting up x axis labels so they don't overlap + no legend
A2SNPcountperTissuegroup

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/A2-Number of SNPs after PheWAS search and LD removal.pdf"), width= 150, height=100, unit="mm")
```

#### A3 - number_of_SNP-trait_associations_per_tissue_group after LD and pipeline run
This plot shows the overview of our SNPs in each Tissue Group from the final df. (i.e after PheWAS search and LD removal but then filtered to remove duplicated Tissue Groups from the trait matching process)
```{r}
LD_Gtex = df%>%
  dplyr::select(Tissue, tissue_group, rsid, p, `P-Value`, NES, position_gtex, chr_gtex, a, b, NES_direction)
LD_Gtex = unique(LD_Gtex)# Removes duplicated tissues

list_of_tissuegroups = unique(Full_tissue_database$tissue_group)
list_of_tissuegroups = sort(list_of_tissuegroups)

test = LD_Gtex %>% group_by(tissue_group, NES_direction) %>% summarise(count = length(tissue_group[!is.na(tissue_group)])) # frequency by direction

LD_Gtex$tissue_group = ifelse(LD_Gtex$tissue_group=="Male_Specific_Tissue", "Male Specific Tissue", LD_Gtex$tissue_group)

# loop to add in the missing tissuegroups
for(tissuegroup in list_of_tissuegroups){
  for(n in 1:nrow(test)){
    if(nrow(test[which(test$tissue_group==tissuegroup),])==0){
    test[nrow(test)+1,]= test[nrow(test)+1,] # make a new row
    test$tissue_group[nrow(test)] =tissuegroup
    test$count[nrow(test)] = 0
    test$NES_direction[nrow(test)] = "Positive" # set it to either, doesn't matter as value is 0
    }
  }
}

test$NES_direction = factor(test$NES_direction, levels=c("Positive", "Negative"))
test= test[ which(is.na(test$tissue_group)==F),]

A3STAcountperTissuegroup = ggplot(test, aes(x=tissue_group, y=count, fill=NES_direction)) + 
  geom_col()+ # bar chart counting how many times a tissue appears
  #ylim(0,0.5)+
  scale_fill_viridis_d(begin=1, end=0, name=paste0("Expression Change\n          (NES)")) + # I like this palette but you can remove it or change it 
  labs(title= "Number of SNP - Trait Associations per Tissue (after LD removal)" , x= "Tissue Group", y=paste("Number of STAs"))+ #labels
  theme(axis.text.x = element_text(angle = 45, hjust =1),
        legend.position = "top",
         panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
    scale_x_discrete(expand = c(0.04, 0)) +
    scale_y_continuous(expand = c(0, 0)) 
  # setting up x axis labels so they don't overlap + no legend
A3STAcountperTissuegroup

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/A3-Number of STAs per Tissue (after LD removal).pdf"), width= 150, height=100, unit="mm")
```

#### A4
G - saved later on

#### B - "correlation_of_traits_to_total_possible_traits" Correlation between nmuber of SNPs in each trait group compared to trait database category size - Checking the pattern isn't just beecause of this:
```{r Plotting the raw Gtex egene data against Tissue line graph}
Full_trait_database = read_csv(paste0(file_directory,"/Trait and Tissue Groups/Full_trait_database.csv"))
Full_trait_database = unique(Full_trait_database)
df_RED = df %>% dplyr::select(trait, traitgroup, p) 
df_RED = unique(df_RED)

df_forcomp = df_RED %>% group_by(traitgroup) %>% summarise(count = length(traitgroup[!is.na(traitgroup)])) %>% dplyr::select(traitgroup, count)

TG_count = Full_trait_database %>% group_by(traitgroup) %>% summarise(count = length(traitgroup[!is.na(traitgroup)]))
names(TG_count) = c("traitgroup", "count_full")

comparison_TG = full_join(df_forcomp, TG_count, by=c("traitgroup"))
comparison_TG =comparison_TG %>% filter(!is.na(traitgroup))
comparison_TG$count = ifelse(is.na(comparison_TG$count), 0, comparison_TG$count)


comparison_TG$count_rel =(comparison_TG$count/sum(comparison_TG$count) )*100
comparison_TG$count_full_rel =(comparison_TG$count_full/sum(comparison_TG$count_full) )*100

rsq <- function (x, y) cor(x, y, method = "pearson") ^ 2
r2 = rsq(comparison_TG[,4],comparison_TG[,5])

comparison_TG$count_rel_dif =Mod(comparison_TG$count_rel-comparison_TG$count_full_rel)

BCorrelation_tg = ggplot(comparison_TG, aes(x=count_rel, y=count_full_rel, color = traitgroup)) + #Plot of SNPs NES vs beta from original dataset
  #geom_jitter(aes(color=NES)) +
  geom_point()+ 
   # colourblind friendly palette
  scale_color_manual(values = grid_col) +
  labs(title= paste0("Correlation of STAs per traitgroup from our df to full database"), x= "% of STAs", y= "% Possible Traits in Database") +
  #theme(legend.position = "none")+ 
  geom_line(inherit.aes = T, data = comparison_TG[,4:5], stat="smooth", method = "lm", formula = y~x, alpha = 0.5, color = "black")+
  scale_y_continuous(expand = c(0.1, 0.1))+
    theme(
        axis.text.x = element_text( size=14),
        #legend.position = "none",
        axis.text.y = element_text(size=14),
        axis.title.y = element_text(  size=16),
        axis.title.x = element_text(  size=16),
        panel.background = element_rect(fill="white"),
          panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
  coord_flip()+ 
  annotate(geom="text",  x=50, y=30, label = paste0("r2 = ", round(r2[1],4)),
              fill="black")+ 
  geom_line(inherit.aes=F,aes(x=`count_full_rel`,y=`count_full_rel`), linetype=2, color="darkorchid2")+
    annotate(geom="text",  x=45, y=60, label = paste0("perfect\nrandom\nchance"),color="darkorchid3")
BCorrelation_tg

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/B1-Correlation.pdf"), width= 200, height=100, unit="mm")

B2 = BCorrelation_tg +xlim(c(0,15))+ylim(c(0,15))
B2

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/B2-zoom.pdf"), width= 200, height=100, unit="mm")
# BCorrelation_tg=B2Correlation_tg+
#   theme(legend.position = "none")+
#   geom_text(data=subset(comparison_TG, count_rel_dif>4) , aes(label=traitgroup), vjust=-0.6, hjust=0.2)
# BCorrelation_tg
# ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/B-correlation_of_traits_to_total_possible_traits.pdf"), width= 100, height=100, unit="mm")


```
#### C - Position_count_OVERVIEW: 
Plot of mutation position against their number of effects on expression with direction
```{r Number of snps per tissue plot USING LINKAGE2, fig.height=14, fig.width=4}
LD_Gtex = df%>%
  dplyr::select(Tissue, tissue_group, position_gtex, rsid, `P-Value`, NES, position_gtex, chr_gtex, a, b, NES_direction)
LD_Gtex = unique(LD_Gtex)
LD_Gtex$position_gtex = as.character(LD_Gtex$position_gtex)
LD_Gtex = filter(LD_Gtex, !is.na(Tissue))
Chromosome = df$chr_phewas[1]
colors_man <- c("Positive" = "#FDE725FF", "Negative" = "#440154FF") 
LD_Gtex$NES_direction=factor(LD_Gtex$NES_direction, levels= c("Positive","Negative"))

CPosition_count_OVERVIEW = ggplot(LD_Gtex, aes(x=position_gtex,  fill = NES_direction)) + 
  geom_bar()+ # bar chart counting number of times rsid appears
  scale_fill_manual(values = colors_man) +
  labs(title= paste0("Number of tissues each SNP\n affects expression in", gene_name) , x= paste0("Position (Chr ",Chromosome, ")"), y = "Number of Tissues")+ #labels
  theme(legend.position = "top",
        #legend.box.spacing = unit(0.1, 'cm'),
        #legend.box.margin = unit(0.1, 'cm'),
        legend.justification = c(1,-1),
        legend.key.width = unit(0.2, 'cm'),
        legend.text = element_text(size=8),
        legend.title = element_text(size=8),
         panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_blank(),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
     scale_x_discrete(expand = c(0, 0)) +
     scale_y_continuous(expand = c(0, 0))+
    coord_flip()

CPosition_count_OVERVIEW

LD_Gtex$position_gtex = as.numeric(LD_Gtex$position_gtex)

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/C-Number of tissues each SNP affects expression in.pdf"), width= 60, unit="mm")
```

####D Overview of trait-SNP associations per traitgroup (incl empty ones) - no direction
```{r}
summary_of_df =df %>% 
  dplyr::select(rsid, traitgroup,trait) %>% 
  group_by(traitgroup) %>%
  summarise(traitgroup_count = n()) %>%
  arrange(traitgroup_count)

all_traitgroups = Full_trait_database %>% 
  dplyr::select(traitgroup, trait) %>%
  mutate(trait =0) %>%
  rename(traitgroup_count = trait) %>%
  unique()
all_traitgroups=all_traitgroups[!(all_traitgroups$traitgroup%in%summary_of_df$traitgroup),]
summary_of_df =full_join(summary_of_df, all_traitgroups)
summary_of_df = summary_of_df %>% filter(!is.na(traitgroup))


DSNP_by_traitgroup= ggplot(summary_of_df, aes(x=reorder(traitgroup, -traitgroup_count), y=traitgroup_count, fill= traitgroup))+
  geom_col()+
  geom_text(check_overlap = T,aes( y=traitgroup_count, label=traitgroup_count), vjust=-0.5)+
  scale_fill_manual(values = grid_col)+
  theme(legend.position = "none",
        axis.text.x = element_text( angle =70, hjust=1),
                 panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
    scale_x_discrete(expand = c(0.04, 0)) +
    scale_y_continuous( expand = expansion(mult = c(0, .1))) +
  labs(x = "Trait Group", y = "Number of SNP-Trait Associations", title = "Number of SNP-Trait Associations in Each Traitgroup")
DSNP_by_traitgroup
ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/D1 Number of STAs in Each Traitgroup.pdf"), width=200,height=120, unit="mm")

#Alternate version, relative to Trait Group size#####


traitgroup_counts = Full_trait_database %>% 
  count(traitgroup)
names(traitgroup_counts) =c("traitgroup", "total_traits_in_database")

summary_of_df = left_join(summary_of_df, traitgroup_counts)
summary_of_df$rel.traitgroup = summary_of_df$traitgroup_count/summary_of_df$total_traits_in_database

D2RELSNP_by_traitgroup= ggplot(summary_of_df, aes(x=reorder(traitgroup, -rel.traitgroup), y=rel.traitgroup, fill= traitgroup))+
  geom_col()+
 # geom_text(check_overlap = T,aes( y=rel.traitgroup, label=rel.traitgroup), vjust=-0.5)+
  scale_fill_manual(values = grid_col)+
  theme(legend.position = "none",
        axis.text.x = element_text( angle =70, hjust=1),
                 panel.background = element_rect(fill="white"),
         panel.grid.major.x = element_blank(),
         panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        axis.line = element_line(colour = "black", linetype="solid", size=1))+
    scale_x_discrete(expand = c(0.04, 0)) +
    scale_y_continuous( expand = expansion(mult = c(0, .1))) +
  labs(x = "Trait Group", y = paste0("STAs per Trait Group/\ntotal traits in database"), title = "Proportion of STAs in Each Traitgroup Relative to Database")
D2RELSNP_by_traitgroup
ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/D2 Proportion of STAs in Each Traitgroup Relative to Database.pdf"), width=200,height=120, unit="mm")

```
#### E - pvalvstraitgroup Raw Number of STAs
#### E1 Proportional count of the STAs in each beta direction and expression direction:
```{r relative counts vs trait overview SEATTLE}
colors_man <- c("Positive" = "#FDE725FF", "Negative" = "#440154FF") 
# Remove duplicates from snps in multiple tissues and split by NES
temp_pos = df %>% dplyr::filter(NES_direction=="Positive") %>%
  dplyr::select(traitgroup, beta_direction, rsid, p , trait, NES_direction)%>%
  unique()
temp_neg = df %>% dplyr::filter(NES_direction=="Negative") %>%
  dplyr::select(traitgroup, beta_direction, rsid, p , trait, NES_direction) %>%
  unique()

E1 = ggplot()+ #Plot of Traitgroup vs SNP Count
  geom_bar(data= temp_pos, aes(x=factor(traitgroup), fill=beta_direction, y=after_stat(count)/nrow(temp_pos)),  position = position_dodge2(width = 1, preserve = "single"))+
  geom_bar(data= temp_neg, aes(x=factor(traitgroup), fill=beta_direction, y=(after_stat(count)/nrow(temp_neg))*-1), position = position_dodge2(width = 1, preserve = "single"))+##  alpha is transparancy from 0-1
    geom_hline(yintercept=0, color ="gray70")+
  #ylim(-0.175,0.175)+
  scale_fill_manual(values = colors_man) +
  labs(title= paste0("Number of STAs per Traitgroup Relative to Total Number of Traitgroups in this Dataframe") , x= "Trait Group", y= paste("Relative Number of STAs (n/Total)"))+ #labels
  theme(axis.text.x = element_text(angle = 60, hjust =1, size = 8),
        axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"),
        legend.position = "top"
        )  + 
  scale_x_discrete(limits=trait_order)

E1  

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/E1 STAs relative to total in df.pdf"), width=200, height =100 , units="mm")



```

#### E2 Proportional (WITHOUT other gene expression change (OGEC))
```{r relative counts NO OGEC vs trait overview SEATTLE, fig.width = 8, fig.height =4}
# Remove duplicates from snps in multiple tissues and split by NES
temp_pos = df %>% dplyr::filter(NES_direction=="Positive"& traitgroup!="Other Gene Expression Change"& traitgroup!="Other/Unspecified") %>%
  dplyr::select(traitgroup, beta_direction, rsid, p , trait, NES_direction)%>%
  unique()
temp_neg = df %>% dplyr::filter(NES_direction=="Negative"& traitgroup!="Other Gene Expression Change"& traitgroup!="Other/Unspecified") %>%
  dplyr::select(traitgroup, beta_direction, rsid, p , trait, NES_direction) %>%
  unique()

E2 = ggplot()+ #Plot of Traitgroup vs SNP Count
  geom_bar(data= temp_pos, aes(x=traitgroup, fill=beta_direction, y=after_stat(count)/nrow(temp_pos)),  position = position_dodge2(width = 1, preserve = "single"))+
  geom_bar(data= temp_neg, aes(x=traitgroup, fill=beta_direction, y=(after_stat(count)/nrow(temp_neg))*-1), position = position_dodge2(width = 1, preserve = "single"))+##  alpha is transparancy from 0-1
    geom_hline(yintercept=0, color ="gray70")+
  scale_fill_manual(values = colors_man) +
  labs(title= paste0("Number of STAs per Traitgroup Relative to Total Number of Traitgroups in this Dataframe") , x= "Trait Group", y= paste("Relative Number of STAs (n/Total)"))+ #labels
  theme(axis.text.x = element_text(angle = 60, hjust =1, size = 8),
        axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"),
        legend.position = "top")  + 
  scale_x_discrete(limits=trait_order[1:22]) 

E2

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/E2-E1 no OGEC or Other.pdf"), width=200, height =100 , units="mm")

```

#### F
#####1 - all
The overview of all snps relative to traitgroups:
INCOMPLETE - ADJUST FONT SIZES FOR FINAL FIGURE!
```{r NO CUTOFF Trait Groups Effected by SNPs that Change in Expression of YOUR_GENE_NAME on Traits by logp}

summary_of_df = summary_of_df %>% dplyr::arrange(-traitgroup_count) %>% mutate(Rank =c(1:nrow(summary_of_df)))

summary_of_df_no_OGEC = summary_of_df %>% filter(traitgroup!="Other Gene Expression Change" &traitgroup!="Other/Unspecified" )%>% dplyr::arrange(-traitgroup_count) %>% mutate(Rank_no_OGEC =c(1:(nrow(summary_of_df)-2)))

temp = df[ which(is.na(df$tissue_group)==F),]
temp = left_join(temp,summary_of_df)
temp$beta_direction = factor(temp$beta_direction, levels=c("Positive", "Negative"))
#Now select only the traits to avoid tissue duplication
temp = temp %>% dplyr::select(trait,traitgroup,logp_DIR,beta_direction,rsid, Rank) %>%
  unique()
temp$traitgroup = factor(temp$traitgroup, levels =trait_order)



F1pvalvstraitgroup = ggplot(temp, aes(x=traitgroup, y=logp_DIR, color=beta_direction)) + #Plot of Traitgroup vs p (P-Value is the Gtex P not PS p)
  geom_point(alpha=0.6, position = position_jitterdodge())+ #  alpha is transparancy from 0-1
  scale_colour_manual(values = colors_man) + # I like this palette but you can remove it or change it - green/blue option begin=0.2, end=0.7
  labs(title= paste0("Overview of STAs in Traitgroups and their -log(p) values") , x= "Trait Group", y= "-log(p) x Direction of expression change")+ #labels
  geom_hline(aes(yintercept=0,), alpha=0.3)+ # faint line at 0
  theme(axis.text.x = element_text(angle = 60, hjust =1, size = 8),
        axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"),
        legend.position = "top")+
  geom_text(inherit.aes=F,aes(label = Rank,x=traitgroup,y=min(logp_DIR), fill="black"), check_overlap = T, vjust=0.5)

  
F1pvalvstraitgroup 


ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/F1 Overview of STAs in Traitgroups and their -log(p) values.pdf"), width=200, height =100 , units="mm")
```


#####F2 Without Other or OGEC - often skews data due to high amount of SNPS in these categories


```{r NO CUTOFF Trait Groups Affected by SNPs that Change in Expression of YOUR_GENE_NAME on Traits by logp}

temp = df[ which(is.na(df$tissue_group)==F),]
temp = temp[which(temp$traitgroup!="Other/Unspecified" &temp$traitgroup!="Other Gene Expression Change" ),]
temp = left_join(temp,summary_of_df_no_OGEC)
temp$beta_direction = factor(temp$beta_direction, levels=c("Positive", "Negative"))
temp = temp %>% dplyr::select(trait,traitgroup,logp_DIR,beta_direction,rsid, Rank_no_OGEC) %>%
  unique()
temp$traitgroup = factor(temp$traitgroup, 
                         levels =trait_order)

F2pvalvstraitgroupNO_O = ggplot(temp, aes(x=traitgroup, y=logp_DIR, color=beta_direction)) + 
  geom_point(alpha=0.6, position = position_jitterdodge())+ #  alpha is transparancy from 0-1
  scale_colour_manual(values = colors_man) + 
  labs(title= paste0("Overview of STAs in Traitgroups (No Other or Other Gene Expression Change)") , x= "Trait Group", y= "-log(p) x Direction of expression change")+ #labels
  geom_hline(aes(yintercept=0,), alpha=0.3)+ # faint line at 0
theme(axis.text.x = element_text(angle = 60, hjust =1, size = 8),
        axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"),
      legend.position = "top"
      )+
    geom_text(inherit.aes=F,aes(label = Rank_no_OGEC,x=traitgroup,y=min(logp_DIR), fill="black"), check_overlap = T, vjust=1.5)+
  scale_y_continuous(expand = c(0.1, 0))

F2pvalvstraitgroupNO_O

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/F2 Overview of STAs in Traitgroups (No Other or Other Gene Expression Change).pdf"), width=200, height =100 , units="mm")

```

#####F3 only Other Gene Expression Change
The overview of all snps relative to traitgroups:

```{r NO CUTOFF Trait Groups Effected by SNPs that Change in Expression of YOUR_GENE_NAME on Traits by logp}
temp = df[ which(is.na(df$tissue_group)==F),]
temp$beta_direction = factor(temp$beta_direction, levels=c("Positive", "Negative"))
temp = temp[which(temp$traitgroup=="Other Gene Expression Change"),]
temp = temp %>% dplyr::select(trait,traitgroup,logp_DIR,beta_direction,rsid) %>%
  unique()
# Clean up as many ENSG's as we can
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- temp %>% dplyr::filter(grepl("ENSG",trait)) 
genes <-genes$trait
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id",
"hgnc_symbol"),values=genes,mart= mart)
G_list = G_list %>% filter(nchar(hgnc_symbol)>0) 
names(G_list)=c("trait","correction")
temp=left_join(temp,G_list)
temp$trait=ifelse(is.na(temp$correction), temp$trait,temp$correction)

F3OGEC = ggplot(temp, aes(x=trait, y=logp_DIR, color=beta_direction)) + 
  geom_point(alpha=0.6, position = position_jitterdodge())+ #  alpha is transparancy from 0-1
  scale_colour_manual(values = colors_man) + 
  labs( x= "Genes", y= "-log(p) * Direction")+ 
  geom_hline(aes(yintercept=0,), alpha=0.3)+ # faint line at 0
  theme(axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.background = element_rect(fill="white"),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
        legend.position = "top")+
  coord_flip()+
      force_panelsizes(cols=unit(1, "in"))#cols=unit(1,"in"))
  
F3OGEC

filename=paste0(file_directory,"/",gene_name,"_test/Final_plots/F3 OGEC only.pdf")

length_of_unique_traits = length(unique(temp$trait))
  if(length_of_unique_traits>150){
    ggsave(filename, height=560, width=230, unit="mm")
  }else if(length_of_unique_traits<150 &length_of_unique_traits>50){
    ggsave(filename, height=380, width=230, unit="mm")
  }else if(length_of_unique_traits<50){
    ggsave(filename, height=250, width=230, unit="mm")
  }else if(length_of_unique_traits<300){
    ggsave(filename, height=130, width=230, unit="mm")
  }

```
#####F3-2 only Other 
The overview of all snps relative to traitgroups:

```{r NO CUTOFF Trait Groups Effected by SNPs that Change in Expression of YOUR_GENE_NAME on Traits by logp}
temp = df[ which(is.na(df$tissue_group)==F),]
temp$beta_direction = factor(temp$beta_direction, levels=c("Positive", "Negative"))
temp = temp[which(temp$traitgroup=="Other/Unspecified"),]
temp = temp %>% dplyr::select(trait,traitgroup,logp_DIR,beta_direction,rsid) %>%
  unique()


F3_O = ggplot(temp, aes(x=trait, y=logp_DIR, color=beta_direction)) + 
  geom_point(alpha=0.6, position = position_jitterdodge())+ #  alpha is transparancy from 0-1
  scale_colour_manual(values = colors_man) + 
  labs( x= "Traits", y= "-log(p) * Direction")+ 
  geom_hline(aes(yintercept=0,), alpha=0.3)+ # faint line at 0
  theme(axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.background = element_rect(fill="white"),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
        legend.position = "top")+
  coord_flip()+
      force_panelsizes(cols=unit(1, "in"))#cols=unit(1,"in"))
  
F3_O

filename=paste0(file_directory,"/",gene_name,"_test/Final_plots/F3-2 Other only.pdf")

length_of_unique_traits = length(unique(temp$trait))
  if(length_of_unique_traits>150){
    ggsave(filename, height=560, width=230, unit="mm")
  }else if(length_of_unique_traits<150 &length_of_unique_traits>50){
    ggsave(filename, height=380, width=230, unit="mm")
  }else if(length_of_unique_traits<50){
    ggsave(filename, height=250, width=230, unit="mm")
  }else if(length_of_unique_traits<300){
    ggsave(filename, height=130, width=230, unit="mm")
  }

```

#### F4 Top n categories after other and OGEC:

```{r, fig.width=6, fig.height=10}
number_of_traitgroups = df %>% dplyr::select(traitgroup) %>% unique()
number_of_traitgroups = number_of_traitgroups %>% 
  dplyr::filter(traitgroup!="Other Gene Expression Change"|
                traitgroup!="Other/Unspecified")

n_tg_for_shiny = nrow(number_of_traitgroups)-2
for(i in 3:nrow(number_of_traitgroups)-2){
  print(i)
  traitgroup_i = summary_of_df_no_OGEC$traitgroup[i]
  temp =df[which(df$traitgroup ==traitgroup_i),] 
  for_label = df[which(df$traitgroup ==traitgroup_i),]
  for_label = length(unique(for_label$trait))
  if(length(unique(temp$Subcat))==1){
  pvsTrait_bygroup = ggplot(temp, aes(x=trait, y=logp_DIR, color=beta_direction)) + 
    geom_point(position = position_jitterdodge(), size =2)+ #  alpha is transparancy from 0-1
    scale_colour_manual(values = colors_man) +
    labs(title= paste0("Traits in ",traitgroup_i," Affected by SNPs that Change in Expression of ", gene_name) , x= "Traits", y= "-log(p) x Direction")+ #labels
    geom_hline(aes(yintercept=0,), alpha=0.3)+# faint line at 0
    geom_hline(aes(yintercept=4,), alpha=0.3, color="red")+
    geom_hline(aes(yintercept=-4,), alpha=0.3, color="red")+
    theme(axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.background = element_rect(fill="white"),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
        legend.position = "top")+
    #facet_grid(rows = vars(Subcat), scales = "free", space="free")+
    coord_flip()+
    force_panelsizes(cols=unit(40, "mm"))
  pvsTrait_bygroup
  }else{
  pvsTrait_bygroup = ggplot(temp, aes(x=trait, y=logp_DIR, color=beta_direction)) + 
    geom_point(position = position_jitterdodge(), size =2)+ #  alpha is transparancy from 0-1
    scale_colour_manual(values = colors_man) +
    labs(title= paste0("Traits in ",traitgroup_i," Affected by SNPs that Change in Expression of ", gene_name) , x= "Traits", y= "-log(p) x Direction")+ #labels
    geom_hline(aes(yintercept=0,), alpha=0.3)+# faint line at 0
    geom_hline(aes(yintercept=4,), alpha=0.3, color="red")+
    geom_hline(aes(yintercept=-4,), alpha=0.3, color="red")+
    theme(axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.background = element_rect(fill="white"),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
        legend.position = "top")+
    facet_grid(rows = vars(Subcat), scales = "free", space="free")+
    coord_flip()+
    force_panelsizes(cols=unit(40, "mm"))
  pvsTrait_bygroup
    }
    
# Manual version:
# ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/E4-Number_", i,"_traitgroup.pdf"),height=y, width=x,  )}
#  sub('f','F',df$address)
  traitgroup_i_ = sub("/","_or_",traitgroup_i)
assign(paste0("F4_", i), pvsTrait_bygroup) 
filename=paste0(file_directory,"/",gene_name,"_test/Final_plots/F4-Number_", i,"_traitgroup",traitgroup_i_,".pdf")
  if(for_label>150){
    ggsave(filename, height=560, width=230, unit="mm")
  }else if(for_label<150 &for_label>50){
    ggsave(filename, height=380, width=230, unit="mm")
  }else if(for_label<50){
    ggsave(filename, height=250, width=230, unit="mm")
  }else if(for_label<25){
    ggsave(filename, height=130, width=230, unit="mm")
  }else if(for_label<10){
    ggsave(filename, height=60, width=230, unit="mm")
  }
}

```


# Part 2 - More plots, focussing on gene region and comparisons
## Firstly it is important to gather information about your gene structure, exon and intron locations using ensembl.
```{r}
# Complicated package, essentially grabbing the most common transcripts, exons, introns and UTRs
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
getAtt = c('chromosome_name', 'start_position', 'end_position',
'strand','exon_chrom_start','exon_chrom_end', '5_utr_start','5_utr_end','3_utr_start','3_utr_end', "ensembl_transcript_id")
gene_structure = getBM(attributes=getAtt,filters=c("hgnc_symbol","with_ccds"),value=list(gene_name,TRUE),mart=ensembl)
gene_structure$hgnc_symbol = gene_name
write_csv(gene_structure,paste0(file_directory,"/",gene_name,"_test/Ensembl/",gene_name,"_structure_FIN.csv"))

```
#### 1. Assigning Exon, intron and UTR to the final_df
```{r}
df = read_csv(paste0(file_directory,"/",gene_name,"_test/Final_data/final_df.csv"))

limitsSTART = min(gene_structure$start_position)# earliest start
limitsEND = max(gene_structure$end_position) # latest finish
# 1st some NA's from the gnomad merge:
df$position_gtex =ifelse(is.na(df$position_gtex), df$Position, df$position_gtex)

# setting up a "within Gene region" column by testing if position is between the limits
## some strands mean gene is running backward in position, in this case it is important to flip the logic of the loop using an if argument:

  df$Within_Gene_Region = "No"
if(limitsSTART>limitsEND){
  for(n in 1:nrow(df)){
    if(limitsSTART>=df$position_gtex[n] & df$position_gtex[n]>=limitsEND){
      df$Within_Gene_Region[n] = "Yes"
    }
  }
}else{
      for(n in 1:nrow(df)){
    if(limitsSTART<=df$position_gtex[n] & df$position_gtex[n]<=limitsEND){
      df$Within_Gene_Region[n] = "Yes"
    }
  }
}

# Now repeated for 5' and 3' UTR: INCOMPLETE - only needed if the next step doesn't do that!

df$Intron_or_Exon_or_UTR = "Intron" 
for( n in 1:nrow(df)){
  for(i in 1:nrow(gene_structure)){
    exon=gene_structure$exon_chrom_start[i]
    exon=append(exon, gene_structure$exon_chrom_end[i])
    prime5=gene_structure$`5_utr_start`[i]
    prime5=append(prime5, gene_structure$`5_utr_end`[i])
    prime3=gene_structure$exon_chrom_start[i]
    prime3=append(prime3, gene_structure$exon_chrom_end[i])
    if(min(exon)<=df$position_gtex[n] & df$position_gtex[n]<=max(exon)){
      df$Intron_or_Exon_or_UTR[n] = "Exon"
    }
    else if(is.na(min(prime5))==F & min(prime5)<=df$position_gtex[n] & df$position_gtex[n]<=max(prime5)){
      df$Intron_or_Exon_or_UTR[n] = "5_UTR"
    }
    else if(is.na(min(prime3))==F & min(prime3)<=df$position_gtex[n] & df$position_gtex[n]<=max(prime3)){
      df$Intron_or_Exon_or_UTR[n] = "3_UTR"
    }
    
  }

}

# setting up None option in intron/exon - overwrites intron with None
for(n in 1:nrow(df)){
  if(df$Within_Gene_Region[n]=="No"){
    df$Intron_or_Exon_or_UTR[n] = "None"
  }
}
unique(df$Intron_or_Exon_or_UTR)

# repeated for Gtex RAW data

Gtex_cleaned = read_csv(paste0(file_directory,"/", gene_name,"_test/Gtex/", gene_name, "_Gtex_cleaned.csv"))
  Gtex_cleaned$Within_Gene_Region = "No"
if(limitsSTART>limitsEND){
  for(n in 1:nrow(Gtex_cleaned)){
    if(limitsSTART>=Gtex_cleaned$position_gtex[n] & Gtex_cleaned$position_gtex[n]>=limitsEND){
      Gtex_cleaned$Within_Gene_Region[n] = "Yes"
    }
  }
}else{
      for(n in 1:nrow(Gtex_cleaned)){
    if(limitsSTART<=Gtex_cleaned$position_gtex[n] & Gtex_cleaned$position_gtex[n]<=limitsEND){
      Gtex_cleaned$Within_Gene_Region[n] = "Yes"
    }
  }
}
  

Gtex_cleaned$Intron_or_Exon_or_UTR = "Intron" 
for( n in 1:nrow(Gtex_cleaned)){
  for(i in 1:nrow(gene_structure)){
    exon=gene_structure$exon_chrom_start[i]
    exon=append(exon, gene_structure$exon_chrom_end[i])
    prime5=gene_structure$`5_utr_start`[i]
    prime5=append(prime5, gene_structure$`5_utr_end`[i])
    prime3=gene_structure$exon_chrom_start[i]
    prime3=append(prime3, gene_structure$exon_chrom_end[i])
    if(min(exon)<=Gtex_cleaned$position_gtex[n] & Gtex_cleaned$position_gtex[n]<=max(exon)){
      Gtex_cleaned$Intron_or_Exon_or_UTR[n] = "Exon"
    }
    else if(is.na(min(prime5))==F & min(prime5)<=Gtex_cleaned$position_gtex[n] & Gtex_cleaned$position_gtex[n]<=max(prime5)){
      Gtex_cleaned$Intron_or_Exon_or_UTR[n] = "5_UTR"
    }
    else if(is.na(min(prime3))==F & min(prime3)<=Gtex_cleaned$position_gtex[n] & Gtex_cleaned$position_gtex[n]<=max(prime3)){
      Gtex_cleaned$Intron_or_Exon_or_UTR[n] = "3_UTR"
    }
    
  }

}

# setting up None option in intron/exon - overwrites intron with None
for(n in 1:nrow(Gtex_cleaned)){
  if(Gtex_cleaned$Within_Gene_Region[n]=="No"){
    Gtex_cleaned$Intron_or_Exon_or_UTR[n] = "None"
  }
}
unique(Gtex_cleaned$Intron_or_Exon_or_UTR)




 # Only run these next 2 lines if it's the first time, merges changes with the previous data file for simplicity
write_csv(df, paste0(file_directory,"/",gene_name,"_test/Final_data/final_df.csv"))
write_csv(Gtex_cleaned, paste0(file_directory,"/",gene_name,"_test/Gtex/Gtex_cleaned.csv"))

```

### Preamble, data setup steps 
#### 2. Regulatory region setup
**I have now created an automated version of this using the biomaRt package**
1. Set Mart to funcgen - (regulatory features) and data to hsapiens_regulatory_feature
2. Next choose chromosome and start/end based on Gtex data and merged
3. Run getBM() to pull out human regulatory features within this region
```{r Assigning regulatory regions}
ensembl = useMart("ENSEMBL_MART_FUNCGEN")
ensembl = useDataset("hsapiens_regulatory_feature",mart=ensembl)
# make varibles for chromosome and start and end + fudge factor (to include final points)
chr = df$chr_phewas[1]
gene_region_st = min(Gtex_cleaned$position_gtex-10000) # 9640000
gene_region_fin = max(Gtex_cleaned$position_gtex+10000) #  11400000
filter_values=list(chr,gene_region_st,gene_region_fin)
temp = getBM(attributes=c("chromosome_name", "chromosome_start", "chromosome_end", "feature_type_name"), 
      filters = c("chromosome_name", "start", "end"),
      values = filter_values, 
      mart = ensembl)

write_csv(temp, paste0(file_directory,"/",gene_name,"_test/Ensembl/",gene_name,"_full_REG_region.csv"))
```
Now we assign those reg features to df and gtex dataframes
```{r}
reg_feat = temp
df$reg_region ="None" # default set to "None"

for(n in 1:nrow(df)){
  for(i in 1:nrow(reg_feat)){
    if(df$position_gtex[n]>= reg_feat$chromosome_start[i] & df$position_gtex[n]<=reg_feat$chromosome_end[i]){ # if our snp is within one of these boundaries
      df$reg_region[n] = reg_feat$feature_type_name[i]
    }
  }
}
#repeated for Gtex
Gtex_cleaned$reg_region ="None" # default set to "None"

for(n in 1:nrow(Gtex_cleaned)){
  for(i in 1:nrow(reg_feat)){
    if(Gtex_cleaned$position_gtex[n]>= reg_feat$chromosome_start[i] & Gtex_cleaned$position_gtex[n]<=reg_feat$chromosome_end[i]){ # if our snp is within one of these boundaries
      Gtex_cleaned$reg_region[n] = reg_feat$feature_type_name[i]
    }
  }
}
write_csv(df, paste0(file_directory,"/",gene_name,"_test/Final_data/final_df.csv"))
write_csv(Gtex_cleaned, paste0(file_directory,"/",gene_name,"_test/Gtex/Gtex_cleaned_REG.csv"))
```

Ok so this next bit is getting more and more complicated... Essentially There are a lot of transcripts. So what is an exon or an intron varies on where you look. ON the plots we will see all possible exons

First load in main dataframe with bound regions and the structure as a dataframe called *rectangles*
```{r load in main dataframe with bound regions and the structure as a dataframe called rectangles, results=FALSE, message=FALSE}

Gtex = read_csv(paste0(file_directory,"/",gene_name,"_test/Gtex/Gtex_cleaned_REG.csv"))
df = read_csv(paste0(file_directory,"/",gene_name,"_test/Final_data/final_df.csv"))
reg_feat = read_csv(paste0(file_directory,"/",gene_name,"_test/Ensembl/",gene_name,"_full_REG_region.csv")) 
gene_rectangles = read_csv(paste0(file_directory,"/",gene_name,"_test/Ensembl/",gene_name,"_structure_FIN.csv")) 
```

Now, importantly, both for the linear and curved gene diagram the leg work remains the same. 
* Define the regions by having start, end and feature columns (as in gene_rectangles). 
* Next define the y axis as the minimum p value to the maximum p value (pmin to pmax) Plot these as rectangles where the introns and exons take up the middle two quadrants and the reg regions fill the rest.
* Overlay your SNPs from the appropriate dataframe with x = position and y = p value  

First, we need to take the regulatory feature dataframe (downloaded form ensembl BioMART - see start of section) and clean it to look like gene_rectangles
```{r Setting up reg feature dataframe }
# Now simplifying the regulatroy region page down
reg_feat_for_bind = reg_feat[,2:4]
# sorting the names to be the same as the gene_rectangles 
reg_feat_for_bind = reg_feat_for_bind %>% rename(Start=chromosome_start,
                                                 End=chromosome_end,
                                                 EI=feature_type_name)

gene_rectangles = gene_rectangles[,5:6]
gene_rectangles$EI ="Exon"
gene_rectangles = gene_rectangles %>% rename(Start=exon_chrom_start,
                                                 End=exon_chrom_end)

# and finally binding the two table together
gene_and_reg = rbind(reg_feat_for_bind, gene_rectangles)

```

### Gene diagram plotting
Next we begin plotting the basic linear setup. I have also included the 'random y value' version if you'd prefer not to consider p-value. **NOTE:** _RED - indicates the reduced SNP set associated with searching the Gtex SNPs in PheWAS and removing SNPS based on LD. _G indicates the original Gtex set of SNPs WITHOUT LD removal.
**NOTE 2:** In the plotting lines there is a legend.position = "none" line. This removes the legend for a higher resolution plot, However it is important to have at least one with the legend for figure building. For this, comment out the line and save as _Legend.  

#### Color setup preamble
This section shifts around a fair bit in terms of which variables are in each plot and I wanted to have a consistent colour scheme throughout. So this chunk serves to set up and save colour information.
```{r}
cols = c("CTCF Binding Site" = "#C39BD3",
         "Enhancer" = "#196F3D70", 
         "Exon" = "#2C3E50", 
         "Intron"= "#56B4E9",
         "None" = "#C0392B",
         "Open chromatin" = "#AAB7B8",
         "Promoter" ="#1ABC9C",
         "Promoter Flanking Region" ="#ABEBC6",
         "TF binding" ="#E67E22"
         )

```


####0. LEGACY, moved to ARCHIVE 

  
####1. GTEX SNPs (all, pre-merge and LD removal)
```{r Linear _G diagram}
# select relevent columns, position, which reg_region the SNP is in, and the P-value of this SNP-Tissue correlation
df_G = Gtex %>% dplyr::select(position_gtex, reg_region, `P-Value`) 
# Setting up a -logP column to make it more readable
df_G$logp=-log10(df_G$`P-Value`)
# df = as.data.frame(df$position_gtex)
df_G=unique(df_G) # Creating a reduced version of the positions of the snps from the final dataframe
df_G = arrange(df_G, position_gtex) # Sort numerically by position
df_G = as.data.frame(df_G) 

# Next is assigning y values (rectangle height) to `P-Value`
pmin = min(df_G$logp) # assign min and max p values
pmax = max(df_G$logp)

# pmax-pmin/4 essentially means the range of p-values divided by 4, i.e. the middle 2 quadrants
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second
# number of snps plotted:
intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(gene_start=min(Start), gene_end=max(End))
if(intron_hack$gene_start > intron_hack$gene_end){
  names(intron_hack)=c("gene_end", "gene_start") # flips names
}
  
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])
intron_hack$EI ="Intron"

Unique_snps = length(unique(df_G$position_gtex))


# Setting up the plot - not commented in great detail
gene_structure_G = ggplot()+
  geom_rect(gene_and_reg[which(gene_and_reg$EI !="Intron" & gene_and_reg$EI !="Exon"),], mapping=aes(xmin=Start, xmax=End, ymin=yminD, ymax=ymaxD,  fill=EI ),  alpha=0.8)+
  geom_rect(data=intron_hack, mapping=aes(xmin=gene_start, xmax=gene_end, ymin=yminD, ymax=ymaxD, fill = EI))+
  geom_rect(gene_and_reg[which(gene_and_reg$EI =="Exon"),], mapping=aes(xmin=Start, xmax=End, ymin=yminD, ymax=ymaxD, color=EI))+
  geom_point(aes(x= df_G$position_gtex, y=df_G$logp, color = df_G$reg_region), size=0.6, alpha =0.6)+ # alpha=0 , add this for _legend plot
  theme(panel.grid.major.x = element_line(size=0.2),
        panel.grid.minor.x = element_line(size=0.05),
        axis.text.x = element_text(angle = 70, hjust =1),
        legend.title = element_blank(),
        legend.position = "none", 
        aspect.ratio = 0.1,
        panel.background = element_rect(fill = "white",
                                        colour = "white"))+
  labs(title= paste0(gene_name, " Gene Region Diagram (Gtex SNPs)") , x= "Position", y= "-log(p)",caption = paste("Overall number of SNPs =", Unique_snps))+
  scale_fill_manual(values=cols)+
  scale_color_manual(values=cols)+
  scale_x_reverse()

gene_structure_G
ggsave(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output/Gene_structure_G.pdf"))


```


####2b. Gtex SNPs (P-Value) after merge and LD removal
```{r df_gtex_RED2}
df_gtex_RED = df %>% dplyr::select(position_gtex, reg_region, `P-Value`)
df_gtex_RED$logp=-log10(df_gtex_RED$`P-Value`)
# df_RED = as.data.frame(df$position_gtex)
df_gtex_RED=unique(df_gtex_RED) # Creating a reduced version of the positions of the snps from the final dataframe
df_gtex_RED = arrange(df_gtex_RED, position_gtex)
df_gtex_RED = as.data.frame(df_gtex_RED)
df_gtex_RED =na.omit(df_gtex_RED)

# Next is assigning y values (rectangle height) to `P-Value`
pmin = min(df_gtex_RED$logp)
pmax = max(df_gtex_RED$logp)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(gene_start=min(Start), gene_end=max(End))
if(intron_hack$gene_start > intron_hack$gene_end){
  names(intron_hack)=c("gene_end", "gene_start") # flips names
}
  
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])
intron_hack$EI ="Intron"

# number of snps plotted:
Unique_snps = length(unique(df_gtex_RED$position_gtex))

#Plot:
gene_structure_RED = ggplot()+
  geom_rect(gene_and_reg[which(gene_and_reg$EI !="Intron" & gene_and_reg$EI !="Exon"),], mapping=aes(xmin=Start, xmax=End, ymin=yminD, ymax=ymaxD,  fill=EI ),  alpha=0.8)+
  geom_rect(data=intron_hack, mapping=aes(xmin=gene_start, xmax=gene_end, ymin=yminD, ymax=ymaxD, fill = EI),  alpha=0.6)+
  geom_rect(gene_and_reg[which(gene_and_reg$EI =="Exon"),], mapping=aes(xmin=Start, xmax=End, ymin=yminD, ymax=ymaxD, color=EI))+
  geom_point(aes(x= df_gtex_RED$position_gtex, y=df_gtex_RED$logp, color = df_gtex_RED$reg_region), size=0.6)+
  theme(panel.grid.major.x = element_line(size=0.2),
        panel.grid.minor.x = element_line(size=0.05),
        axis.text.x = element_text(angle = 70, hjust =1),
        legend.title = element_blank(),
        legend.position = "none",
        aspect.ratio = 0.1,
        panel.background = element_rect(fill = "white",
                                        colour = "white"))+
  labs(title= paste0(gene_name," Gene Region Diagram (Gtex after Merge and LD removal SNPs)") , x= "Position", y= "-log(p)",caption = paste("Overall number of SNPs =", Unique_snps))+
  scale_fill_manual(values=cols)+
  scale_color_manual(values=cols)+
  scale_x_reverse()




gene_structure_RED



ggsave(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output/Gene_structure_gtex_RED.pdf"))
```



####4. Half-circular Gene diagrams (For Circos Figure)
This Section has been moved before the Circos proper because we are still making gene diagrams using only the data we have setup in this "Gene Diagram" section. However for an explanation of circlize and tutorials on how to use it, please do read the next section "Circos".

All of these are also available with SNPs coloured based on the reg_region they are in (uncomment those lines), Also, change pch from 20 to 21 for rings as opposed to points which can look better, 
##### Curved Gtex (all, pre-merge and LD removal)
```{r Curved Gene diagrams Gtex_G, error=TRUE}

setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))

# Similar to above, select relevent columns, remove duplicates and 



df_for_circle = Gtex %>% dplyr::select(chr_gtex, position_gtex, reg_region, `P-Value`)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$`P-Value`)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]

gene_and_reg =gene_and_reg %>% filter(EI!="Intron") %>% dplyr::select(Start,End,EI,yminD,ymaxD)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(Start=min(Start), End=max(End))
if(intron_hack$Start > intron_hack$End){
  names(intron_hack)=c("End", "Start") # flips names
}

intron_hack$EI ="Intron" 
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])

gene_and_reg= rbind(gene_and_reg,intron_hack)

gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

# colours_for_rect = c("palegreen3","gray70","violet","cyan3","gold","red3","black","skyblue1")

gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))

circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_G.pdf") # add _color to name to add color for snps


circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               # These 3 lines are alter the outer label (in this case chrNUMBER)
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               # These lines are really important and WILL need modifying to get it looking good on the circle, digits = number of significant figures
               # by = is tick rate
               circos.axis(labels.cex = 0.4, 
                           direction ="outside", 
                           major.at = seq(signif(df_for_circle$position_gtex[which.min(df_for_circle$position_gtex)], digits = 4),                                                                            signif(df_for_circle$position_gtex[which.max(df_for_circle$position_gtex)], digits = 4), 
                                          by = 50000)) # specifically this "by=" will have to be changed based on how crowded it looks
               # Thes alter teh y axis ticks and will again need altering
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="left", 
                            at = seq(from = as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     length = 4))
             }) 

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3, col = "mediumblue") # add in col=df_for_circle$color

dev.off()
```

##### Curved Gtex (all, pre-merge and LD removal) + Colour
```{r Curved Gene diagrams Gtex_G, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))

# Similar to above, select relevent columns, remove duplicates and 
df_for_circle = Gtex %>% dplyr::select(chr_gtex, position_gtex, reg_region, `P-Value`)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$`P-Value`)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]

gene_and_reg =gene_and_reg %>% filter(EI!="Intron") %>% dplyr::select(Start,End,EI,yminD,ymaxD)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(Start=min(Start), End=max(End))
if(intron_hack$Start > intron_hack$End){
  names(intron_hack)=c("End", "Start") # flips names
}

intron_hack$EI ="Intron" 
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])

gene_and_reg= rbind(gene_and_reg,intron_hack)

gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

# colours_for_rect = c("palegreen3","gray70","violet","cyan3","gold","red3","black","skyblue1")

gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))
# Add in colours for snps:
df_for_circle$color = ifelse( df_for_circle$reg_region== "Enhancer", "#196F3D",
                             ifelse(df_for_circle$reg_region== "Open chromatin", "#AAB7B8",
                                     ifelse(df_for_circle$reg_region== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(df_for_circle$reg_region== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(df_for_circle$reg_region== "TF binding", "#E67E22",
                                                             ifelse(df_for_circle$reg_region== "Promoter", "#1ABC9C",
                                                                     ifelse(df_for_circle$reg_region== "Exon", "#2C3E50",
                                                                           ifelse(df_for_circle$reg_region== "Intron", "#56B4E9","#C0392B"))))))))


# Define rectangles dataframe for the coloured ring underneath the plot - OPTIONAL
range = signif(max(df_for_circle$position_gtex),4)+50000 - signif(min(df_for_circle$position_gtex),4)
scale_bar = seq(from =signif(min(df_for_circle$position_gtex),4) , to = signif(max(df_for_circle$position_gtex),4)+50000 , by= range/10 )
scale_bar =as.data.frame(scale_bar)
names(scale_bar) = "x"
scale_bar$colour = "#9430c2" # - purple
scale_bar$colour[seq(1, nrow(scale_bar), 2)]="#bd680d" # fills every other row with orange

circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_G_Colour.pdf") # add _color to name to add color for snps


circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               # These 3 lines are alter the outer label (in this case chrNUMBER)
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               # These lines are really important and WILL need modifying to get it looking good on the circle, digits = number of significant figures
               # by = is tick rate
               circos.axis(labels.cex = 0.4, 
                           direction ="outside", 
                           major.at = seq(signif(df_for_circle$position_gtex[which.min(df_for_circle$position_gtex)], digits = 4),                                                                            signif(df_for_circle$position_gtex[which.max(df_for_circle$position_gtex)], digits = 4), 
                                          by = 50000)) # specifically this "by=" will have to be changed based on how crowded it looks
               # These alter the y axis ticks and will again need altering
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="left", 
                            at = seq(from = as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     length = 4))
             }) 

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3, col=df_for_circle$color) # add in col=df_for_circle$color
circos.rect(xleft=scale_bar$x[1:nrow(scale_bar)-1], CELL_META$cell.ylim[1] - mm_y(1.5), xright=scale_bar$x[2:nrow(scale_bar)], ytop=CELL_META$cell.ylim[1], col=scale_bar$colour, border=NA) #0-((max(df_for_circle$logp))/10)

dev.off()
```
##### CURVED Gtex SNPs (P-Value) after merge and LD removal
```{r Curved Gene diagrams Gtex_RED, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))

# Similar to above, select relevent columns, remove duplicates and 
df_for_circle = df %>% dplyr::select(chr_gtex, position_gtex, reg_region, `P-Value`)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$`P-Value`)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]

gene_and_reg =gene_and_reg %>% filter(EI!="Intron") %>% dplyr::select(Start,End,EI,yminD,ymaxD)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(Start=min(Start), End=max(End))
if(intron_hack$Start > intron_hack$End){
  names(intron_hack)=c("End", "Start") # flips names
}

intron_hack$EI ="Intron" 
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])

gene_and_reg= rbind(gene_and_reg,intron_hack)

gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second


# colours_for_rect = c("palegreen3","gray70","violet","cyan3","gold","red3","black","skyblue1")

gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))



circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_RED.pdf") # add in _color if using coloured snps


circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               circos.axis(labels.cex = 0.4, direction ="outside", major.at = seq(signif(df_for_circle$position_gtex[which.min(df_for_circle$position_gtex)]-500000, digits = 4),
                                                                                  signif(df_for_circle$position_gtex[which.max(df_for_circle$position_gtex)], digits = 4), 
                                                                                  by = 50000))
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="left", 
                            at = seq(from = as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     length = 4))
             })

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3, col= "mediumblue") # add in col=df_for_circle$color

dev.off()
```
##### CURVED Gtex SNPs (P-Value) after merge and LD removal + Colour
This plot also has the purple and orange split for figure building - commented out for the moment - remove from release -INCOMPLETE:
```{r Curved Gene diagrams Gtex_RED}
setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))

# Similar to above, select relevent columns, remove duplicates and 
df_for_circle = df %>% dplyr::select(chr_gtex, position_gtex, reg_region, `P-Value`)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$`P-Value`)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]
# colours_for_rect = c("palegreen3","gray70","violet","cyan3","gold","red3","black","skyblue1")
gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))

# # Add in colours for snps:
df_for_circle$color = ifelse( df_for_circle$reg_region== "Enhancer", "#196F3D",
                             ifelse(df_for_circle$reg_region== "Open chromatin", "#AAB7B8",
                                     ifelse(df_for_circle$reg_region== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(df_for_circle$reg_region== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(df_for_circle$reg_region== "TF binding", "#E67E22",
                                                             ifelse(df_for_circle$reg_region== "Promoter", "#1ABC9C",
                                                                     ifelse(df_for_circle$reg_region== "Exon", "#2C3E50",
                                                                           ifelse(df_for_circle$reg_region== "Intron", "#56B4E9","#C0392B"))))))))

# Define rectangles dataframe for the coloured ring underneath the plot - OPTIONAL
range = signif(max(df_for_circle$position_gtex),4)+50000 - signif(min(df_for_circle$position_gtex),4)
scale_bar = seq(from =signif(min(df_for_circle$position_gtex),4) , to = signif(max(df_for_circle$position_gtex),4)+50000 , by= range/10 )
scale_bar =as.data.frame(scale_bar)
names(scale_bar) = "x"
scale_bar$colour = "#9430c2" # - purple
scale_bar$colour[seq(1, nrow(scale_bar), 2)]="#bd680d" # fills every other row with orange

### Getting started with circos
circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_RED_colour.pdf") # add in _color if using coloured snps


circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               circos.axis(labels.cex = 0.4, direction ="outside", major.at = seq(signif(df_for_circle$position_gtex[which.min(df_for_circle$position_gtex)]-500000, digits = 4),
                                                                                  signif(df_for_circle$position_gtex[which.max(df_for_circle$position_gtex)], digits = 4), 
                                                                                  by = 50000))
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="left", 
                            at = seq(from = as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     length = 4))
             }) 

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3, col=df_for_circle$color) # add in col=df_for_circle$color

# Adding in the scale bar - Find absolute values for y
circos.rect(xleft=scale_bar$x[1:nrow(scale_bar)-1], CELL_META$cell.ylim[1] - mm_y(1.5), xright=scale_bar$x[2:nrow(scale_bar)], ytop=CELL_META$cell.ylim[1], col=scale_bar$colour, border=NA) #0-((max(df_for_circle$logp))/10)
#circos.rect(xleft=scale_bar$x[1:nrow(scale_bar)-1], ybottom=0-((max(df_for_circle$logp))/3), xright=scale_bar$x[2:nrow(scale_bar)], ytop=0-((max(df_for_circle$logp))/10), col=scale_bar$colour, border=NA) #0-((max(df_for_circle$logp))/10)

dev.off()
```
##### CURVED PheWAS SNPs (p Value) after merge and LD removal
```{r Curved Gene diagrams PHEWAS_RED}
setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))

# Similar to above, select relevent columns, remove duplicates and 
df_for_circle = df %>% dplyr::select(chr_gtex, position_gtex,reg_region, p)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$p)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]
# This dataset has one incredibly significant SNP so option to remove it for nicer plot here:
#df_for_circle = df_for_circle[which(df_for_circle$logp<400),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]

gene_and_reg =gene_and_reg %>% filter(EI!="Intron") %>% dplyr::select(Start,End,EI,yminD,ymaxD)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(Start=min(Start), End=max(End))
if(intron_hack$Start > intron_hack$End){
  names(intron_hack)=c("End", "Start") # flips names
}

intron_hack$EI ="Intron" 
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])

gene_and_reg= rbind(gene_and_reg,intron_hack)

gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second


gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))

circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_P_FULL.pdf") # change to _outlier if outlier is removed line3104; and _color if using coloured snps
circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex, xlim =c(min(gene_and_reg$Start), max(gene_and_reg$End))) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               # These 3 lines are alter the outer label (in this case chrNUMBER)
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               # These lines are really important and WILL need modifying to get it looking good on the circle, digits = number of significant figures
               # 
               circos.axis(labels.cex = 0.4, 
                           direction ="outside", 
                           major.at = seq(signif(gene_and_reg$Start[which.min(gene_and_reg$Start)], digits = 3), 
                                          signif(gene_and_reg$End[which.max(gene_and_reg$End)], digits = 3), 
                                          by = 100000)) 
               # These alter the y axis ticks and will again need altering
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="right", 
                            at = seq(from =as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     by = as.integer(max(df_for_circle$logp)/4)))
             }) 

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3, col= "mediumblue") # add in col=df_for_circle$color 
circos.rect(xleft=scale_bar$x[1:nrow(scale_bar)-1], ybottom=min(df_for_circle$logp)-((max(df_for_circle$logp))/3), xright=scale_bar$x[2:nrow(scale_bar)], ytop=min(df_for_circle$logp)-((max(df_for_circle$logp))/5), col=scale_bar$colour, border=NA) #0-((max(df_for_circle$logp))/10)

dev.off()
```
##### CURVED PheWAS SNPs (p Value) after merge and LD removal + Colour
```{r Curved Gene diagrams PHEWAS_RED}
setwd(paste0(file_directory,"/",gene_name,"_test/Gene_diagram_output"))
# Similar to above, select relevent columns, remove duplicates and 
df_for_circle = df %>% dplyr::select(chr_gtex, position_gtex,reg_region, p)
df_for_circle = unique(df_for_circle)
df_for_circle$y = runif(nrow(df_for_circle)) # random y value between 0 and 1 - stddist
df_for_circle$logp = -log(df_for_circle$p)
df_for_circle = df_for_circle[which(complete.cases(df_for_circle)==T),]
# This dataset has one incredibly significant SNP so option to remove it for nicer plot here:
#df_for_circle = df_for_circle[which(df_for_circle$logp<400),]

pmax=df_for_circle$logp[which.max(df_for_circle$logp)]
pmin=df_for_circle$logp[which.min(df_for_circle$logp)]

gene_and_reg =gene_and_reg %>% filter(EI!="Intron") %>% dplyr::select(Start,End,EI,yminD,ymaxD)
gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

intron_hack = gene_and_reg %>% 
  filter(gene_and_reg$EI =="Exon") %>% 
  summarise(Start=min(Start), End=max(End))
if(intron_hack$Start > intron_hack$End){
  names(intron_hack)=c("End", "Start") # flips names
}

intron_hack$EI ="Intron" 
intron_hack$yminD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,4])
intron_hack$ymaxD = as.numeric(gene_and_reg[which(gene_and_reg$EI =="Exon"),][1,5])

gene_and_reg= rbind(gene_and_reg,intron_hack)

gene_and_reg$yminD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmin+((pmax-pmin)/4),pmin) # setting up arbitrary y values for the rectangles
gene_and_reg$ymaxD = if_else(gene_and_reg$EI =="Intron" | gene_and_reg$EI =="Exon", pmax-((pmax-pmin)/4),pmax)# if it is an intron use the first value, anything else the second

# colours_for_rect = c("palegreen3","gray70","violet","cyan3","gold","red3","black","skyblue1")

gene_and_reg$color = ifelse( gene_and_reg$EI== "Enhancer", "#196F3D", 
                             ifelse(gene_and_reg$EI== "Open chromatin", "#AAB7B8",
                                     ifelse(gene_and_reg$EI== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(gene_and_reg$EI== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(gene_and_reg$EI== "TF binding", "#E67E22",
                                                             ifelse(gene_and_reg$EI== "Promoter", "#1ABC9C",
                                                                     ifelse(gene_and_reg$EI== "Exon", "#2C3E50",
                                                                           ifelse(gene_and_reg$EI== "Intron", "#56B4E9","FAIL"))))))))

# Add in colours for snps:
df_for_circle$color = ifelse( df_for_circle$reg_region== "Enhancer", "#196F3D",
                             ifelse(df_for_circle$reg_region== "Open chromatin", "#AAB7B8",
                                     ifelse(df_for_circle$reg_region== "Promoter Flanking Region", "#ABEBC6",
                                             ifelse(df_for_circle$reg_region== "CTCF Binding Site", "#C39BD3",
                                                     ifelse(df_for_circle$reg_region== "TF binding", "#E67E22",
                                                             ifelse(df_for_circle$reg_region== "Promoter", "#1ABC9C",
                                                                     ifelse(df_for_circle$reg_region== "Exon", "#2C3E50",
                                                                           ifelse(df_for_circle$reg_region== "Intron", "#56B4E9","#C0392B"))))))))

circos.clear()
# .par() start and gap degree for setting partial rings up!

pdf("Curved_gene_diagram_P_FULL_Colour.pdf") # change to _outlier if outlier is removed line3104; and _color if using coloured snps


circos.par("track.height" = 0.1, start.degree = 180 , gap.degree= 180, canvas.xlim =c(-1,1), canvas.ylim =c(-0.3,1)) # radius of all circles = 1 so this means 1/10 is each track
# Track 1 - Gene Information
circos.initialize(df_for_circle$chr_gtex, x = df_for_circle$position_gtex, xlim =c(min(gene_and_reg$Start), max(gene_and_reg$End))) # defining the x axis for the tracks
# next we add in a track
circos.track(df_for_circle$chr_gtex, y = df_for_circle$logp, 
             panel.fun = function(x, y) {
               # These 3 lines are alter the outer label (in this case chrNUMBER)
               circos.text(CELL_META$xcenter, CELL_META$cell.ylim[1] + mm_y(15),
                           CELL_META$sector.index, facing = "clockwise", niceFacing = TRUE,
                           adj = c(0, 0.5), cex = 0.6)
               # These lines are really important and WILL need modifying to get it looking good on the circle, digits = number of significant figures
               # 
               circos.axis(labels.cex = 0.4, 
                           direction ="outside", 
                           major.at = seq(signif(gene_and_reg$Start[which.min(gene_and_reg$Start)], digits = 3), 
                                          signif(gene_and_reg$End[which.max(gene_and_reg$End)], digits = 3), 
                                          by = 100000)) 
               # These alter the y axis ticks and will again need altering
               circos.yaxis(labels.cex=0.4, 
                            labels.niceFacing = TRUE, side="right", 
                            at = seq(from = as.integer(min(df_for_circle$logp)), 
                                     to = as.integer(max(df_for_circle$logp)), 
                                     by = as.integer(max(df_for_circle$logp)/4)))
             }) 

circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$Start, ybottom=pmin, xright=gene_and_reg[which(gene_and_reg$EI != "Exon" & gene_and_reg$EI != "Intron"),]$End, ytop=pmax, col=gene_and_reg$color, border=NA)
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Intron"),]$End, ytop=pmax-0.25*(pmax-pmin), col="#56B4E9", border="#56B4E9")
circos.rect(xleft=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$Start, ybottom=pmin+0.25*(pmax-pmin), xright=gene_and_reg[which(gene_and_reg$EI == "Exon"),]$End, ytop=pmax-0.25*(pmax-pmin), col="black", border="black")
circos.trackPoints(sectors=df_for_circle$chr_gtex, x=df_for_circle$position_gtex,y= df_for_circle$logp, pch = 25, cex = 0.3,col=df_for_circle$color) # add in col=df_for_circle$color 
circos.rect(xleft=scale_bar$x[1:nrow(scale_bar)-1], CELL_META$cell.ylim[1] - mm_y(1.5), xright=scale_bar$x[2:nrow(scale_bar)], ytop=CELL_META$cell.ylim[1], col=scale_bar$colour, border=NA) #0-((max(df_for_circle$logp))/10)

dev.off()
```


####5. Graphs of where the SNPS are: (quick look before circos)
Next I'd like to plot the  bar chart of each reg/genetic element as a basic visualisation of what's going on.
```{r bar plot _G_P bar chart of reg region count}
Unique_snps = df %>% dplyr::select( rsid, Intron_or_Exon_or_UTR, reg_region)
Unique_snps = unique(Unique_snps)
Unique_snps$reg_region = factor(Unique_snps$reg_region,levels = c("CTCF Binding Site", "Enhancer","Open chromatin","TF binding", "Promoter", "Promoter Flanking Region", "None"))

snpvsregion = ggplot(Unique_snps, aes(x = reg_region, fill = Intron_or_Exon_or_UTR))+
  geom_bar(alpha=0.8, position = position_dodge2(width = 1, preserve = "single"))+#  alpha is transparancy from 0-1
  scale_fill_viridis_d(direction=1, option = "magma", end=0.6, name ="Intron, Exon or Neither") + # I like this palette but you can remove it or change it 
  labs(title= "Number of Unique SNPs in Each Type of Genomic Region" , x= "Regulatory Region", y= "Count", caption = paste("Overall number of SNPs =", nrow(Unique_snps)))+ #labels
  theme(axis.text.x = element_text(angle = 45, hjust =1, size = 10),
        axis.text.y =   element_text(size = 12),
        axis.title = element_text(size = 12, face="bold"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"))
snpvsregion




ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/A4-number_of_snps_per_regulatory_region.pdf"))
```



### Circos
This section will not be nearly as automated due to the individuality gene to gene. It is a little difficult to get started so here are some examples with "circlize" package. There is a very comprehensive e-book on this package: https://jokergoo.github.io/circlize_book/book/ that has been invaluable to getting these plots working.

Step 1, load in the dataframes you want. In my case that's the full merged df and the Gtex original df. This will allow me to look at links between the position, tissue group, trait group before and after merge. - most are already in , commented in case you need it.


```{r for circos data load in}
df=df %>% filter(!is.na(tissue_group))
# saving what we have so far in case circos breaks
save.image(paste0(file_directory,"/", gene_name,"_test/Final_data/final_workspace_for_shiny.RData"))
```

Step 2, Now we are going to go through and plot each circos plot first. NOTE: this section uses the generic df name "test" as the frequency table that each circos plot is made from. This means each chunk will override this test table. If you would like to save it then use writetsv(test, "filename.tsv"). But they are quick and easy to make and not much point commiting to memory I've found.

I've also included a chunk with the list format for grid_col, I use a html colour picker to grab colours I want but you can also use circlize's inbuilt col_mat = rand_color(length(test)) (meaning assign random colours for the length of that generic frequency matrix "test") COLOUR STORAGE:

NOTE: The issue here is circlize isn't really cut out for assigning specific colours to lots of things to be consistent between plots. So this colour chunk below is my over-the-top way of assigning specific colours to EVERYTHING. This means that for every new gene I make a new grid_col dataframe, compare grid_col and the list of traitgroups/tissuesgroups for the new gene and see if I need to add any and which I need to comment out.

```{r Storage for Colour}
# Correcting  df  names to actually be plottable
df$traitgroup =ifelse(df$traitgroup =="Muscle/Tendon Traits", "Muscular/Tendon Traits", df$traitgroup)
# FUll colour list
grid_col = c(`Body Fat/Size` = "#0DF485",
             `Diet` = "#39FFC0",
            `Metabolism (inc. Diabetes)` ="#32DCC1",
             `Tumours (Benign/Cancer)` ="#32CCDC" ,
             `Atrophy/Dystrophy` = "#0DB5F4",
            `Congenital Defects` = "#3277DC",
             `Cardiovascular Traits` = "#4532DC",
             `Blood/Vascular Traits` ="#6C39FF",
             Inflammation ="#9A32DC", #  or #AA98A970
            `Respiratory` ="#C432DC",
             `Renal/Urinary` = "#DC32D0",
            `Physical Activity` = "#DC324D",
             `Digestive Traits` ="#DC4132",
            `Nervous System/Brain` = "#DC5D32",
             `Bone Traits (inc. Dental)` = "#DC8732",
            `Muscle/Tendon Traits` = "#DCA732",
            `Physical Injury` = "#DCD232",
            `Alcohol-related Disease` ="#BCDC32",
             `Life Span` ="#92DC32",
             `Female Specific Traits` = "#43DC32",
            `Male Specific Traits` = "#00f556",
            `Disease (unknown origin)` ="#00f577",
            `Other/Unspecified` ="#C9C7BD",
            `Other Gene Expression Change` ="#78C47D",
             #Tissues:
             Blood = "#CC5500",
             `Cultured Cells` = "#00FFFF",
             Digestive ="#A7C7E7",
             Endocrine = "#FF4433",
             Fat = "#E0115F",
             `Female Reproductive Tissue` = "#0C5F7D",
             Heart = "#80461B",
             Lymphatic ="#023020",
             `Male Specific Tissue` ="#FFA500",
             Mammary = "#E9846E",
             Muscle = "#FFBF00",
             Nervous = "#90EE90",
             `Respiratory Tissue` = "#D673C9",
             Skin = "#00A36C",
            Urniary = "darkorchid4",
             Vascular ="#630330"
             
)

# selects just the names from grid_col - Tissuegroup/traitgroup
colour_data = names(grid_col)
# selects tissue/tratgroups from main df
new_groups = append(unique(df$traitgroup), unique(df$tissue_group))
print(new_groups)
# 
to_add = setdiff(new_groups, colour_data) # data unique to colour data
to_comment_out = setdiff(colour_data, new_groups)

```
Now we have two lists, one that tells which to add to grid_col as new groups (to_add) and the other telling us which to comment out (to_comment_out)
```{r}
print(to_add)
```

```{r}
print(to_comment_out)
# removes members of grid_col that overlap with the to comment out
grid_col = grid_col[!(names(grid_col) %in% to_comment_out)]

if(length(colour_data)-length(grid_col)==length(to_comment_out)){
  print("Removed all unnecersary colours as expected")
} else{ print("Something went wrong!")}
```

Now we make a new modified grid_col, copied from above to be modified. then simply put a '#' in front of each line you don't need:
```{r}

grid_col_trait = grid_col[1:length(unique(df$traitgroup))]
grid_col_trait = grid_col_trait[which(!is.na(grid_col_trait))]

grid_col_tissue = grid_col[(length(unique(df$traitgroup))+1):(length(unique(df$traitgroup))+length(unique(df$tissue_group)))]
grid_col_tissue = grid_col_tissue[which(!is.na(grid_col_tissue))]
```

INCOMPETE APRIL23 - Need to add in a caveat for the circos plots - maybe without labels and anually add or something else clever, for the moment leaving it as is and it will probably error


####1. Gtex (from merged df)
```{r Gtex from merged df circos 2, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Circos"))

# First selecting key columns from the Gtex dataframe (pre-merge)
 simple_LUT_Gmerge = df %>% dplyr::select(position_gtex, Tissue, tissue_group, `P-Value`)
 

 # removing any duplicated values
 simple_LUT_Gmerge = unique(simple_LUT_Gmerge)
 # Now simplifying again to just position and tissue_group
 simple_LUT_Gmerge = simple_LUT_Gmerge %>% dplyr::select(position_gtex, tissue_group)
# table() returns a frequency matrix with position down y and tissue_group along x in this case named generically "test"
 test = table(simple_LUT_Gmerge)

# circlize works as follows : 1. initialise the environment (not needed for chordDiagram (links)); Plot track; plot extras ontop of track; add labels; PLot the next track in; etc... In this case we use chordDiagram to plot the links, then circos.track() for the labels. Circos.track is a generic call that allows you to define a function using panel.fun{}. Beyond my scope of knowledge so this is lifted directly from the e-book example
circos.clear() # generic: "Clear everything from the plotting environment" command
pdf("Position_Gtex_P.pdf")
circos.par(start.degree=-90) # Rotates the plot so tissues are on the right as they will in the final figure
chordDiagram(test, 
              annotationTrackHeight = c(0.01, 0.01), 
              annotationTrack = "grid",
              big.gap =10,
              small.gap = 0,
              column.col = grid_col_tissue, # or col_mat
              grid.col = grid_col_tissue,
              diffHeight = mm_h(1), 
              # (length(unique(simple_LUT$tissue_group)))
              preAllocateTracks = list(track.height = 1*max(strwidth(unlist(dimnames(test))))),
              transparency = 0.5
 )
#This particular plot has SO many positions it makes the labels look a bit rubbish but I've kept them in anyway
circos.track(track.index = 1, panel.fun = function(x, y) {
   circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index,
               facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =0.7)
}, bg.border = NA)
dev.off()

for_LUT_plot = simple_LUT_Gmerge %>%
  mutate(arb=1) %>%
  dplyr::select(-position_gtex) %>%
  unique()

for_LUT_plot %>% 
ggplot() +
  geom_col(aes(x=tissue_group, y=arb, fill=tissue_group))+
  scale_fill_manual(values = grid_col)+
  scale_x_discrete(limits=rev,position = "top")+
  coord_flip()+

    theme(axis.text.x =  element_blank(),
          axis.ticks =  element_blank(),
        axis.text.y =   element_text(size = 10),
        axis.title =  element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.y = element_line(colour = "gray90", linetype="solid", size=0.2),
        panel.background = element_rect(fill="white"),
        legend.position = "none")
ggsave(paste0(file_directory,"/",gene_name,"_test/Circos/tissue_LUT.pdf"), width = 70, height=100, unit="mm")

```

####3. Tissue-Traitgroup overlap: NOTE: For these next two the traitgroups are listed here with colours associated. Every gene will have a different list of traitgroups. Best to go through manually and modify accordingly, or use rand_color (see commented lines)
```{r tissue trait group overlap for circos 3, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Circos"))
# This one Tissuegroup-traitgroup - uising the simple_LUT_COR

# OR using rand_color:
# col_mat = rand_color(length(test))
# col_mat = rand_color((2*length(unique(simple_LUT_COR$tissue_group))))
 
simple_LUT_COR = df %>% dplyr::select(tissue_group, traitgroup)


simple_LUT_COR$traitgroup = factor(simple_LUT_COR$traitgroup, levels = names(grid_col_trait))
simple_LUT_COR$tissue_group = factor(simple_LUT_COR$tissue_group, levels =names(grid_col_tissue))

test5 = table(simple_LUT_COR)

circos.clear()
pdf("G-Tissue_trait_overlap.pdf")
circos.par( start.degree = -90 ) 
chordDiagram(test5, 
             annotationTrackHeight = c(0.01, 0.01), 
             annotationTrack = "grid",
             big.gap =10,
            #small.gap = 1,
             column.col = grid_col, # or col_mat
            grid.col = grid_col,
             diffHeight = mm_h(1), 
              # (length(unique(simple_LUT$tissue_group)))
             preAllocateTracks = list(track.height = 1*max(strwidth(unlist(dimnames(test5))))),
            transparency = 0.5,
            order = c(rev(rownames(test5)), colnames(test5)) # sets order (reverse order for tissues for nice plot)
            )
circos.track(track.index = 1, panel.fun = function(x, y) {
  circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
              facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =0.7)
}, bg.border = NA)
dev.off()
```

####3b. Tissue-Traitgroup overlap NO OTHER: 
```{r tissue trait group overlap for circos 3b, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Circos"))

temp = df[which(df$traitgroup!= "Other/Unspecified"),]
 
simple_LUT_COR = temp %>% dplyr::select(tissue_group, traitgroup)


simple_LUT_COR$traitgroup = factor(simple_LUT_COR$traitgroup, levels = names(grid_col_trait))
simple_LUT_COR$tissue_group = factor(simple_LUT_COR$tissue_group, levels =names(grid_col_tissue))

test2 = table(simple_LUT_COR)

circos.clear()
pdf("Tissue_trait_overlap_NO_OTHER.pdf")
circos.par( start.degree = -90 ) 
chordDiagram(test2, 
             annotationTrackHeight = c(0.01, 0.01), 
             annotationTrack = "grid",
             big.gap =15,
            #small.gap = 1,
             column.col = grid_col, # or col_mat
            grid.col = grid_col,
             diffHeight = mm_h(1), 
              # (length(unique(simple_LUT$tissue_group)))
             preAllocateTracks = list(track.height = 1*max(strwidth(unlist(dimnames(test2))))),
            transparency = 0.5,
            order = c(rev(rownames(test2)), colnames(test2)) # sets order (reverse order for tissues for nice plot)
            )
circos.track(track.index = 1, panel.fun = function(x, y) {
  circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
              facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =0.7)
}, bg.border = NA)
dev.off()
```

####4. Traitgroup to Position overlap
```{r traitggroup to position overlap circos 4, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Circos"))

simple_LUT_P = df %>% dplyr::select(position_gtex, traitgroup, p)


simple_LUT_P$traitgroup = factor(simple_LUT_P$traitgroup, levels = names(grid_col_trait))

simple_LUT_P = unique(simple_LUT_P)
simple_LUT_P = simple_LUT_P %>% dplyr::select(position_gtex, traitgroup)

test4 = table(simple_LUT_P)
 
layout(matrix(1:6, 2, 3))
circos.clear()
pdf("trait_position.pdf")
circos.par( start.degree = 90 ) 
chordDiagram(test4, 
              annotationTrackHeight = c(0.01, 0.01), 
              annotationTrack = "grid",
              big.gap =2,
              small.gap = 1,
              column.col = grid_col_trait, # or col_mat
              grid.col = grid_col_trait,
              diffHeight = mm_h(1), 
              # (length(unique(simple_LUT$tissue_group)))
              preAllocateTracks = list(track.height = 1*max(strwidth(unlist(dimnames(test4))))),
              transparency = 0.5,
              order = c(rownames(test4), rev(colnames(test4)))
 )
 circos.track(track.index = 1, panel.fun = function(x, y) {
   circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
               facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =0.5)
 }, bg.border = NA)

dev.off()

```

####4b.Traitgroup_pos overlap NO OTHER: 
```{r tissue trait group overlap for circos 4b, error=TRUE}
setwd(paste0(file_directory,"/",gene_name,"_test/Circos"))
# This one Tissuegroup-traitgroup - uising the simple_LUT_COR

temp = df[which(df$traitgroup!= "Other/Unspecified"),]
 
simple_LUT_P = temp %>% dplyr::select(position_gtex, traitgroup, p)


simple_LUT_P$traitgroup = factor(simple_LUT_P$traitgroup, levels = names(grid_col_trait))

simple_LUT_P = unique(simple_LUT_P)
simple_LUT_P = simple_LUT_P %>% dplyr::select(position_gtex, traitgroup)

test3 = table(simple_LUT_P)
 

circos.clear()
pdf("trait_position_NO_OTHER.pdf")
circos.par( start.degree = 90 ) 
chordDiagram(test3, 
              annotationTrackHeight = c(0.01, 0.01), 
              annotationTrack = "grid",
              big.gap =10,
              #small.gap = 1,
              column.col = grid_col_trait, # or col_mat
              grid.col = grid_col_trait,
              diffHeight = mm_h(1), 
              # (length(unique(simple_LUT$tissue_group)))
              preAllocateTracks = list(track.height = 1*max(strwidth(unlist(dimnames(test3))))),
              transparency = 0.5,
              order = c(rownames(test3), rev(colnames(test3)))
 )
 circos.track(track.index = 1, panel.fun = function(x, y) {
   circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
               facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex =0.5)
 }, bg.border = NA)

dev.off()

```

# Figure of all my favourite plots:
 Get all of the code from the plots , ad in "e1 = "etc and then call them all here
 
 e0snptrsait overview
 e1
 e2
 e3
 e4-EXCLUDED, plot size too tempromental e4-1 , 2, 3 and 4
 
 f1
 f2
 (others in file to peruse)
 INCOMPLETE AUG 23
Can I make a nicer summary sheet, single pdf? is it even worth it?
```{r, include=F, eval=F}

pvalvstraitgroupNO_O_e2_leg = pvalvstraitgroupNO_O_e2+ theme(legend.position = "right")
plot_grid(A2SNPcountperTissuegroup,DSNP_by_traitgroup,pvalvstraitgroupe1,pvalvstraitgroupNO_O_e2_leg, nrow=4, ncol=1, align="v")
ggsave(paste0(file_directory,"/",gene_name,"_test/Final_Plots/Summary_1.pdf"), height=600, width=300, unit="mm")
```
#Final clean up of the final_df to make it as usable as possible straight out of the box

```{r}
df= read_csv(paste0(file_directory,"/", gene_name,"_test/Final_data/final_df.csv"))
df_b=df
col_order <- c("Gene Symbol",
               "rsid",
               "trait",
               "traitgroup",
               "Subcat",                
               "beta_cor",
               "p",
               "logp_DIR",
               "ea",
               "nea",
               "se",
               "n",
               "eaf",
               "id",
               "P-Value",
               "a",
               "b",
               "NES",
               "Tissue_nice",
               "tissue_group",
               "tissue_group_secondary",
               "Within_Gene_Region",
               "Intron_or_Exon_or_UTR",
               "reg_region",
               "|NES|",
               "Tissue" ,
               "Alt_Subcat",
               "beta",
               "AT",
               "chr_phewas",
               "position_phewas",
               "Variant Id",
               "chr_gtex",
               "position_gtex",
               "build_gtex",
               "Gencode Id" ,
               "na_count",
               "beta_direction" ,
               "NES_direction",
               "logp",
               "test"
               )

setdiff(names(df),names(df_b))
df <- df_b[, col_order]
df = unique(df)
write_csv(df,paste0(file_directory,"/", gene_name,"_test/Final_data/final_df.csv"))

```

# Shiny app
```{r}
# List all objects in the environment
all_objects <- ls()

# Define the prefix
prefix <- "F4_"

# Filter objects with the specified prefix
matching_objects <- all_objects[grep(paste0("^", prefix), all_objects)]

# Create a list of the matching objects
object_list <- lapply(matching_objects, get)

# Create a list of labels based on the number of matching objects
label_list <- sprintf("Number %d Trait Group", seq_along(object_list))

# Combine the label_list and object_list to create the final list
my_final_list <- setNames(object_list, label_list)
Top5_choice = c(my_final_list,list("Other Gene Expression Change" = F3OGEC,
                                   "Other" = F3_O))

save.image(paste0(file_directory,"/", gene_name,"_test/Final_data/final_workspace_for_shiny.RData"))
runApp( paste0(file_directory,"/app_with_circos.R"))
```

# Shiny app rerun 
Fill in these 3 again
```{r, include=F}

gene_name = "Gene_name"

file_directory = "file_directory"

library(shiny)
load(paste0(file_directory,"/", gene_name,"_test/Final_data/final_workspace_for_shiny.RData"))
runApp( paste0(file_directory,"/app_with_circos.R"))

```

# Manual version of Number_x_traitgroup
```{r, include=FALSE, eval=F}
number_of_traitgroups = df %>% dplyr::select(traitgroup) %>% unique()
number_of_traitgroups = number_of_traitgroups %>% dplyr::filter(traitgroup!="Other Gene Expression Change"|traitgroup!="Other/Unspecified")


  temp =df[which(df$traitgroup =="traitgroup name"),] 
  for_label = df[which(df$traitgroup =="traitgroup name"),]
  for_label = length(unique(for_label$trait))

  pvsTrait_bygroup = ggplot(temp, aes(x=trait, y=logp_DIR, color=beta_direction)) + 
    geom_point(position = position_jitterdodge(), size =2)+ #  alpha is transparancy from 0-1
    scale_colour_viridis_d(direction=-1, ) +
    labs(title= paste0("Traits in",summary_of_df_no_OGEC$traitgroup[i],  paste0("Affected by SNPs that Change in Expression of ", gene_name)) , x= "Traits", y= "-log(p) x Direction")+ #labels
    geom_hline(aes(yintercept=0,), alpha=0.3)+# faint line at 0
    geom_hline(aes(yintercept=4,), alpha=0.3, color="red")+
    geom_hline(aes(yintercept=-4,), alpha=0.3, color="red")+
    theme(        axis.text.y =   element_text(size = 10),
        axis.title = element_text(size = 10, face="bold"),
        panel.background = element_rect(fill="white"),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_line(colour = "gray90", linetype="solid", size=0.4),
        panel.grid.minor.x = element_line(colour = "gray90", linetype="solid", size=0.2),
        legend.position = "none")+
    facet_grid(rows = vars(Subcat), scales = "free", space="free")+
    coord_flip()+
    force_panelsizes(cols=unit(40, "mm"))
  pvsTrait_bygroup
  

ggsave(paste0(file_directory,"/",gene_name,"_test/Final_plots/E4-Number_TRAITGROUP_NAME_traitgroup.pdf"),height=y, width=x,  )
```
# Extra comments
A lot of this code is probably less efficient than it could be. In general I've found it difficult to find the correct vectorised functions to quickly do what I want. For this reason I find myself leaning heavily on more generic computing logic using 'for' and 'if' loops. This does work but please be patient with certain chunks as they may take a little while to run. Any tips and improvements are very welcome. Please do email me :) . 


